<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Juan Domingo" />


<title>scellpam</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">

div.csl-bib-body { }
div.csl-entry {
clear: both;
}
.hanging div.csl-entry {
margin-left:2em;
text-indent:-2em;
}
div.csl-left-margin {
min-width:2em;
float:left;
}
div.csl-right-inline {
margin-left:2em;
padding-left:1em;
}
div.csl-indent {
margin-left: 2em;
}
</style>

<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">scellpam</h1>
<h4 class="author">Juan Domingo</h4>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scellpam)</span></code></pre></div>
<div id="citation" class="section level1">
<h1>Citation</h1>
<p>To cite this package please, use preferently the paper of BMC
mentioned at the bibliography below with citation <span class="citation">Domingo, Leon, and Dura (2023)</span></p>
</div>
<div id="purpose" class="section level1">
<h1>Purpose</h1>
<p>The package <code>scellpam</code> is meant as a way to apply the PAM
algorithm to results of single-cell RNA-Seq data sets.</p>
<p>Differently to other packages, it can deal with a big number of cells
(limited in principle only by the amount of available RAM) and make the
calculations in parallel.</p>
<p>It uses a special format to store binary matrices on disk, the
<code>jmatrix</code> format. Please, make yourself familiar with the
<code>jmatrix</code> creation and manipulation and with the PAM
application looking first at the other vignettes of this package,
<code>jmatrixsc</code> and <code>parallelpamsc</code>.</p>
<p>WARNING: you must NOT load explicitly neither <code>jmatrix</code>
nor <code>parallelpam</code>. Indeed, you do not need even to install
them. All their functions have been included here, too, so doing
<code>library(scellpam)</code> is enough.</p>
</div>
<div id="workflow" class="section level1">
<h1>Workflow</h1>
<div id="debug-messages" class="section level2">
<h2>Debug messages</h2>
<p>First of all, the package can show quite informative (but sometimes
verbose) messages in the console. To turn on/off such messages you can
use,</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Initially, state of debug is FALSE. Turn it on with</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ScellpamSetDebug</span>(<span class="cn">TRUE</span>,<span class="at">debparpam=</span><span class="cn">TRUE</span>,<span class="at">debjmat=</span><span class="cn">TRUE</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Debugging for scellpam (biological part) of the package set to ON.</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Debugging for parallelpam inside scellpam package set to ON.</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Debugging for jmatrix inside scellpam package set to ON.</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We have also turned on debugging of the parallel PAM algorithm and of the binary matrix creation/manipulation</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># but if this annoys you their default value is FALSE for both cases even when the scellpam debug is set to true, so just do</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ScellpamSetDebug(TRUE)</span></span></code></pre></div>
</div>
<div id="data-loadstorage" class="section level2">
<h2>Data load/storage</h2>
<p>As stated before, the limitations in memory suggested us to use
intermediate binary files stored in disk as a way to keep the results of
each step of the process. These files have a concrete internal format;
see the vignette <code>jmatrixsc</code> in this package to know more
about the <code>jmatrix</code> (<span class="citation">Domingo
(2023a)</span>) package.</p>
<p>Thus, the first step is to load raw data (sequencing counts) coming
either from a .csv file or from the main formats used in single cell
RNA-Seq data sets and store them as a binary matrix file in
<code>jmatrix</code> format.</p>
<p>The input formats can be a <code>.csv</code> file, a sparse matrix
stored into the S4 object created by the Seurat package (<span class="citation">Hao et al. (2021)</span>) or a R NumericMatrix
extracted from any package that uses a single-cell experiment object.
Unfortunately, the internal structure of such objects does not seem to
be the same for all packages so we have not been able to provide a
single function to extract from all of them.</p>
<div id="readwrite-.csv-files" class="section level3">
<h3>Read/write .csv files</h3>
<p>Assuming a <code>.csv</code> input file, the following call reads the
file and writes to disk its binary representation. The
<code>jmatrix</code> format can create full, sparse or symmetric
matrices.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToJMat</span>(<span class="st">&quot;countsfile.csv&quot;</span>,<span class="st">&quot;countsfile.bin&quot;</span>)</span></code></pre></div>
<p>By default, data are stored as a sparse matrix of numbers as floats
(4 bytes per item in most architectures). Indeed, default parameters
are</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToJMat</span>(<span class="st">&quot;countsfile.csv&quot;</span>,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">mtype=</span><span class="st">&quot;sparse&quot;</span>,<span class="at">csep=</span><span class="st">&quot;,&quot;</span>,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">ctype=</span><span class="st">&quot;raw&quot;</span>,<span class="at">valuetype=</span><span class="st">&quot;float&quot;</span>,<span class="at">transpose=</span><span class="cn">FALSE</span>,<span class="at">comment=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>where mtype can be <code>sparse</code> or <code>full</code> and
valuetype can be <code>float</code> or <code>double</code>. Changing the
default values for these parameters is strongly discouraged, except when
the sparse matrix is bigger than the full matrix, which happens when
there are few zeros. In such a case use <code>mtype=&quot;full&quot;</code>. The
names for rows and columns are read from the <code>.csv</code> file and
stored as metadata in the binary file after the raw data of counts.</p>
<p>Other parameters you may (and need to) change are the separation
character for fields in the <code>.csv</code> file (csep) if your fields
are not comma-separated; indeed, you can read tab-separated files (.tsv)
using <code>csep=&#39;\t&#39;</code>. Also, you can set the type of
normalization (ctype) which can be <code>raw</code> to write the raw
value of the counts, <code>log1</code> to write the
<code>log2(counts+1)</code>, <code>rawn</code> which is as
<code>row</code> but normalized and <code>log1n</code> which is like
<code>log1</code> but normalized.</p>
<p>WARNING: normalize ALWAYS NORMALIZES BY COLUMNS (before
transposition, if requested to transpose). The logarithm is taken
base-2.</p>
<p>Finally, you may want to write the matrix transposed and add any
comment of up to 1024 characters as a string, if you want. Such comment
will be stored in the metadata, after the row and column names.</p>
<p>In our case, the data are organized with genes by rows and cells by
columns (that is why we choose to normalize by colums), which is not
what is needed later to calculate the dissimilarity matrix. Therefore,
our typical call would be</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToJMat</span>(<span class="st">&quot;countsfile.csv&quot;</span>,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">ctype=</span><span class="st">&quot;log1n&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">comment=</span><span class="st">&quot;Obtained from countsfile.csv&quot;</span>)</span></code></pre></div>
<p>As a result of the calculations the functions of this package
generate binary files, also in <code>jmatrix</code> format. If you want
to inspect them they can be written back as <code>.csv</code> files,
which allows them to be imported as R tables, if your computer has
memory enough to hold them. This is done as</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatToCsv</span>(<span class="st">&quot;countsfile.bin&quot;</span>,<span class="st">&quot;countsback.csv&quot;</span>,<span class="at">csep=</span><span class="st">&quot;,&quot;</span>,<span class="at">withquotes=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>Also, you can selectively read some rows and/or columns of them. See
appropriate functions in the vignette called <code>jmatrixsc</code>
attached to this package.</p>
<p>Notice that in this case <code>counstfile.csv</code> and
<code>countsback.csv</code> would not be equal; row and column names
will have been swapped and that is OK, since we wrote the binary file
with <code>transpose=TRUE</code>.</p>
<p>Also, remember that if you use comma as the separator character, you
can just get rid of the csep parameter. The parameter withquotes, if you
decide to use it and you set it to TRUE, writes each row and column name
in the <code>.csv</code> surrounded by double quotes (but not the
numbers).</p>
<p>It is fair to have <code>.csv</code> files since they can be
inspected to check errors but obviously these files can be quite large
and writing them may be quite slow.</p>
</div>
<div id="readwrite-seurat-dgcmatrix-files" class="section level3">
<h3>Read/write Seurat (dgCMatrix) files</h3>
<p>If your data are not in a <code>.csv</code> file but stored in a S4
object created by package <code>seurat</code> (<span class="citation">Hao et al. (2021)</span>) (and frequently stored for
example in a compressed <code>.rds</code> file) it is likely that this
is an object of the class <code>dgCMatrix</code>, defined in package
<code>Matrix</code> (<span class="citation">Bates, Maechler, and Jagan
(2022)</span>). We provide a function, too, to write it to
<code>jmatrix</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;yourdata.rds&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dgCMatToJMat</span>(q<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>counts,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">comment=</span><span class="st">&quot;Obtained from yourdata.rds&quot;</span>)</span></code></pre></div>
<p>Nevertheless, sometimes the dgCMatrix is stored in a different slot
of the object like</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dgCMatToJMat</span>(q<span class="sc">@</span>raw.data,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">comment=</span><span class="st">&quot;Obtained from other data&quot;</span>)</span></code></pre></div>
<p>It is up to you to inspect the original object (with
<code>str(q)</code>) and see what you must do.</p>
<p>Parameters <code>mtype</code>, <code>ctype</code> and
<code>valuetype</code> are available, too, with the same meaning as for
the former function. The <code>.bin</code> file can obviously be dumped
to <code>.csv</code>, as explained before.</p>
<p>Also, in the case of Seurat, each cell belongs to a sample, being the
sample a set of cells with some common characteristic like coming from
the same living being or from a common stage in a process of cellular
change. Cells are then labelled with a factor and this could be obtained
with</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>Gr<span class="ot">=</span><span class="fu">GetSeuratGroups</span>(q)</span></code></pre></div>
<p>where <code>Gr</code> is a numeric vector of integers with as many
components as cells that contains a consecutive numeric identifier of
the group starting at 1. But again this may depend on the internal
structure of that particular Seurat object.</p>
</div>
<div id="readwrite-from-other-single-cell-experiment-packages" class="section level3">
<h3>Read/write from other single cell experiment packages</h3>
<p>Similarly, if you are using packages that use the <code>sce</code>
(Single-cell experiment) S4 object (like <code>splatter</code> (<span class="citation">Zappia, Phipson, and Oshlack (2017)</span>),
<code>DuoClustering</code> (<span class="citation">Duò and Soneson
(2021)</span>) and others) you will have to extract the matrix of
counts. Unfortunately, again this does not seem to be uniform since each
package has a slightly different internal structure for the S4 object.
The most we have been able to do is to provide a function which takes
the sparse matrix of counts (as a numeric matrix) and the name of the
generated binary file (as before) and optionally the row and column
names (as <code>StringVectors</code>) if they were not attached to the
matrix itself, followed by the parameters used for the csv/Seurat
formats.</p>
<p>The next examples are not marked for execution since they need the
loading of packages splatter/scatter and DuoClustering2018 respectively
and a test to see if they are available using installed.packages is
forbidden if you want to pass the CRAN tests. It you want to run them,
please copy+paste this code in R.</p>
<p>Your code, for example with <code>splatter</code>, would probably
look like</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">library</span>(splatter)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">library</span>(scater)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>Splattersce <span class="ot">&lt;-</span> <span class="fu">mockSCE</span>()</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="fu">SceToJMat</span>(Splattersce<span class="sc">@</span>assays<span class="sc">@</span>data<span class="sc">@</span>listData<span class="sc">$</span>counts,<span class="st">&quot;mockSCE.bin&quot;</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">mtype=</span><span class="st">&quot;full&quot;</span>,<span class="at">ctype=</span><span class="st">&quot;log1n&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">comment=</span><span class="st">&quot;Generated by splatter with mockSCE() and normalized</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="st">                    to log2(counts+1).&quot;</span>)</span></code></pre></div>
<p>Nevertheless, in <code>DuoClustering</code> you would have something
like</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">library</span>(DuoClustering2018)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>KuTCCsce <span class="ot">&lt;-</span> <span class="fu">sce_filteredExpr10_KumarTCC</span>()</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">SceToJMat</span>(KuTCCsce<span class="sc">@</span>assays<span class="sc">$</span>data<span class="sc">@</span>listData<span class="sc">$</span>counts,<span class="st">&quot;KuTCC.bin&quot;</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">mtype=</span><span class="st">&quot;full&quot;</span>,<span class="at">ctype=</span><span class="st">&quot;log1n&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">comment=</span><span class="st">&quot;Generated by DuoCLustering with Kumar data and</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="st">                normalized to log2(counts+1).&quot;</span>)</span></code></pre></div>
<p>Please, notice the subtle difference between</p>
<p><code>Splattersce@assays@data@listData$counts</code> and</p>
<p><code>KuTCCsce@assays$data@listData$counts</code></p>
<p>We have noticed that in simulations with <code>splatter</code>, and
depending on the parameters, the number of zeros can be not too high and
therefore storing the binary matrix as sparse matrix makes its size
bigger, instead of smaller. In such a case you can store it as a full
matrix passing the parameter <code>mtype=&quot;full&quot;</code> to
<code>SceToJMat</code>.</p>
<p>Parameters <code>ctype</code> and <code>valuetype</code> are
available, too, with the same meaning as for the former functions.</p>
</div>
<div id="getting-information-on-the-stored-matrix" class="section level3">
<h3>Getting information on the stored matrix</h3>
<p>Other useful function in this step is <code>JMatInfo</code>, which
prints in the console or in a text file information about a stored
binary file. As an example, let us load and write a <code>.csv</code>
file and then let us see the information of the generated binary
file.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>tmpfile<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToJMat</span>(<span class="st">&quot;extdata/conquer_GSE52529_Trapnell_sample.csv&quot;</span>,tmpfile,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">transpose=</span><span class="cn">TRUE</span>,<span class="at">comment=</span><span class="st">&quot;Experiment conquer GSE52529-GPL16791&quot;</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 288 columns of values (not including the column of names) in file extdata/conquer_GSE52529_Trapnell_sample.csv.</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5218 lines (excluding header) in file extdata/conquer_GSE52529_Trapnell_sample.csv</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data will be read from each line and stored as float values.</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading line... 0 1000 2000 3000 4000 5000 </span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read 5218 data lines of file extdata/conquer_GSE52529_Trapnell_sample.csv, as expected.</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Transposing matrix of (5218x288) to a matrix of (288x5218)</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/Rtmpao5EW9/Trapnell.bin of (288x5218)</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 3030456</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (288 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing column names (5218 strings written, from ENSG00000164675.10 to ENSG00000093100.13).</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Experiment conquer GSE52529-GPL16791</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatInfo</span>(tmpfile)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; File:               /tmp/Rtmpao5EW9/Trapnell.bin</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Matrix type:        SparseMatrix</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of elements: 1502784</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data type:          float</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Endianness:         little endian (same as this machine)</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of rows:     288</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of columns:  5218</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata:           Stored names of rows and columns.</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata comment:  &quot;Experiment conquer GSE52529-GPL16791&quot;</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Binary data size:   3030328 bytes, which is 50.4119% of the full matrix size (which would be 6011136 bytes).</span></span></code></pre></div>
<p>If you want to keep the information about the binary file in a text
file, just do</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>tmpfile<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>tmptxtfile<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellInfo.txt&quot;</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatInfo</span>(tmpfile,tmptxtfile)</span></code></pre></div>
</div>
<div id="filtering-genes-or-cells" class="section level3">
<h3>Filtering genes or cells</h3>
<p>The binary files can be expurged getting rid of some genes or cells.
The function provided for this task is <code>FilterBinByName</code>,
which takes as inputs a binary file and a R list of R-strings with the
names of the genes or cells that must remain (first and second
parameters). The third one is the name of the binary filtered file,
which will have the same nature (full or sparse matrix) and data type of
the matrix in the original file. The fourth parameter,
<code>namesat</code>, indicates if whatever you want to filter (genes or
cells) are in columns (“cols”) or in rows (“rows”). As an example, let
us filter some 1000 of the 65218 genes contained in file
conquer_<code>GSE52529-GPL16791_Trapnell.csv</code> and whose names we
have put in the list stored in variable <code>v</code> by reading the
file <code>Trapnell_remain_genes.csv</code>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>v<span class="ot">&lt;-</span><span class="fu">as.vector</span>(<span class="fu">read.table</span>(<span class="st">&quot;extdata/Trapnell_remain_genes.csv&quot;</span>)[[<span class="dv">1</span>]])</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>tmpfiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell_filtered.bin&quot;</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">FilterJMatByName</span>(tmpfile1,v,tmpfiltfile1,<span class="at">namesat=</span><span class="st">&quot;cols&quot;</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1000 columns of the 5218 in the original matrix will be kept.</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/Rtmpao5EW9/Trapnell_filtered.bin of (288x1000)</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 1276144</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (288 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing column names (1000 strings written, from ENSG00000067113.16 to ENSG00000093100.13).</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Experiment conquer GSE52529-GPL16791</span></span></code></pre></div>
<p>If the list contains names not present in the original file the
program emits a warning (and keeps only whose which are present).</p>
</div>
</div>
<div id="calculating-the-distancedissimilarity-matrix" class="section level2">
<h2>Calculating the distance/dissimilarity matrix</h2>
<p>This is the most computationally intensive part of the process
(particularly, for samples with a high number of cells) and therefore
has been programmed in parallel, taking advantage of the multiple cores
of the machine, if available. See vignette <code>parallelpamsc</code> of
this package to know more about <code>parallelpam</code> (<span class="citation">Domingo (2023b)</span>). From now on, just the
prospective call to comment on the parameters.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">CalcAndWriteDissimilarityMatrix</span>(tmpfile1,tmppeafile1,<span class="at">distype=</span><span class="st">&quot;Pearson&quot;</span>,<span class="at">nthreads=</span><span class="dv">0</span>,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">comment=</span><span class="st">&quot;Pearson dissimilarities for cell in experiment conquer GSE52529-GPL16791&quot;</span>)</span></code></pre></div>
<p>The input and output files (first and second parameters) are of
course compulsory. Input file can be a sparse of full binary matrix (but
obviously, not a symmetric matrix).</p>
<p>WARNING: notice that the vectors to calculate dissimilarities amongst
them MUST be stored BY ROWS. This is due to efficiency reasons and it is
the reason by which we transposed the counts matrix before writing it as
binary.</p>
<p>Output dissimilarity matrix will always be a binary symmetric
(obviously square) matrix with a number of rows (and columns) equal to
the number of rows (in this case, cells) of the counts file. The type of
distance/dissimilarity can be <code>L1</code> (Manhattan distance),
<code>L2</code> (Euclidean distance) or <code>Pearson</code> (Pearson
dissimilarity coefficient). The resulting matrix stores only names for
the rows, which are the names of the cells stored as rows in file
<code>Trapnell.bin</code>. If the number of cells is <span class="math inline">\(N\)</span>, only <span class="math inline">\(N(N+1)/2\)</span> dissimilarity values are really
stored.</p>
<p>A note on the number of threads, valid also for other algorithms that
will be explained later. Possible values for the number of threads
is:</p>
<ul>
<li><p><code>-1</code> (or any negative number) to indicate you do not
want to use threads (strictly sequential computation).</p></li>
<li><p><code>0</code> to allow the program to choose the number of
threads according to the problem size and the number of available
cores.</p></li>
<li><p>Any positive number to force the use of such number of
threads.</p></li>
</ul>
<p>Choosing explicitly a number of threads bigger than the number of
available<br />
cores is allowed, but discouraged and the program emits a warning about
it.</p>
<p>With respect to option <code>0</code> (the program chooses), for
small problems (in this case, less than 1000 cells) the function makes
the choice of not using threads, since the overhead of opening and
waiting termination is not worth. For bigger problems the number of
chosen threads is the number of available cores. Nevertheless, this
choice may not be the best, depending on your machine.</p>
<p>Now, let us really do the call with this small dataset.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">CalcAndWriteDissimilarityMatrix</span>(tmpfile1,tmppeafile1,<span class="at">nthreads=</span><span class="sc">-</span><span class="dv">1</span>,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">comment=</span><span class="st">&quot;Pearson dissimilarities for cell</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="st">                                         in experiment conquer</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="st">                                         GSE52529-GPL16791&quot;</span>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Input matrix is a sparse matrix  with elements of type &#39;float&#39; and size (288,5218)</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read sparse matrix from file /tmp/Rtmpao5EW9/Trapnell.bin.</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Its size is [288 x 5218] and it uses 378647 elements, half of 4 bytes and half of 4 bytes each, with accounts for 2.88995 MBytes.</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Creating dissimilarity matrix of size (288x288)</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 255967224 Kib.</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of dissimilarity matrix calculation (serial version). Elapsed time: 0.728399 s</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/Rtmpao5EW9/TrapnellPearson.bin</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 166592</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (288 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Pearson dissimilarities for cell</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          in experiment conquer</span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          GSE52529-GPL16791</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Output binary file /tmp/Rtmpao5EW9/TrapnellPearson.bin written.</span></span></code></pre></div>
<p>WARNING: the normal way of calling
<code>CalcAndWriteDissimilarityMatrix</code> would use nthreads=0 to
make use of all available cores in your machine. Nevertheless, this does
not seem to be allowed by CRAN to pass the test so I have had to use the
serial version invoked with nthreads=-1. In your normal use of code try
always nthreads=0.</p>
<p>The resulting matrix is stored as a binary symmetric matrix of float
values, as we can check.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatInfo</span>(tmppeafile1)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; File:               /tmp/Rtmpao5EW9/TrapnellPearson.bin</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Matrix type:        SymmetricMatrix</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of elements: 82944 (41616 really stored)</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data type:          float</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Endianness:         little endian (same as this machine)</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of rows:     288</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of columns:  288</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata:           Stored only names of rows.</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata comment:  &quot;Pearson dissimilarities for cell</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          in experiment conquer</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          GSE52529-GPL16791&quot;</span></span></code></pre></div>
</div>
<div id="applying-pam" class="section level2">
<h2>Applying PAM</h2>
<p>The last step is to take the previously calculated matrix and apply
the Partitioning Around Medoids classifier. Function is
<code>ApplyPAM</code>. Default parameters are:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>L<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;BUILD&quot;</span>,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">max_iter=</span><span class="dv">1000</span>,<span class="at">nthreads=</span><span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/Rtmpao5EW9/TrapnellPearson.bin</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 255994140 Kib.</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting BUILD initialization method, serial version</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 50. TD=11914.134766</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 28. 8 reassigned points. TD=10251.591797</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 284. 79 reassigned points. TD=9409.731445</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 107. 0 reassigned points. TD=8996.875977</span></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 268. 21 reassigned points. TD=8661.666016</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 118. 0 reassigned points. TD=8333.292969</span></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 17. 108 reassigned points. TD=8016.939453</span></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 154. 0 reassigned points. TD=7747.802734</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 126. 0 reassigned points. TD=7500.490234</span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 173. 5 reassigned points. TD=7287.086914</span></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 7287.086914</span></span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; BUILD initialization method (serial version) finished. Elapsed time: 0.00187303 s</span></span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting improved FastPAM1 method in serial implementation...</span></span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 0. Medoid at place 1 (point 51) swapped with point 209; TD-change=-103.419434; TD=7183.665039. 60 reassigned points.</span></span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 1. Medoid at place 7 (point 17) swapped with point 39; TD-change=-43.391033; TD=7140.273926. 11 reassigned points.</span></span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 2.    Exiting, since DeltaTDst is 0.000000. Final value of TD is 7140.273926</span></span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Optimization method (serial version) finished. Elapsed time: 0.000909435 s</span></span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.00187303 s (method BUILD).</span></span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0.000909435 s in 1 iterations (0.000909435 seconds/iteration).</span></span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00278247 s (0 minutes, 0.00278247 seconds).</span></span></code></pre></div>
<p>WARNING: the normal way of calling <code>ApplyPAM</code> would use
nthreads=0 to make use of all available cores in your machine.
Nevertheless, this does not seem to be allowed by CRAN to pass the test
so I have had to use the serial version invoked with nthreads=-1. In
your normal use of code try always nthreads=0.</p>
<p>The dissimilarity matrix, as formerly calculated, is compulsory, as
long as the number of medoids, <code>k</code>.</p>
<p>Parameters <code>init_method</code> (and one optional parameter,
<code>initial_med</code>) deserve special comment. The first is the
method to initialize the medoids. Its possible values are
<code>BUILD</code>, <code>LAB</code> and <code>PREV</code>. The rest of
the algorithm make medoid swapping between the points of the initial set
made with <code>BUILD</code> or <code>LAB</code> and the rest of points
until no swap can reduce the objective function, which is the sum of
distances of each point to its closest medoid. But this may fall (and
indeed falls) in local minima. If you initialize with <code>BUILD</code>
or <code>LAB</code> the optional parameter <code>initial_med</code>
cannot be used.</p>
<p>The initialization methods <code>BUILD</code> and <code>LAB</code>
are described in the paper by Schubert at al. (<span class="citation">Schubert and Rousseeuw (2019)</span>).
<code>BUILD</code> is deterministic. <code>LAB</code> uses a sample of
the total points to initialize. Obviously, you can run <code>LAB</code>
to get different initializations and compare the results.</p>
<p>The returned object is a list with two fields: <code>med</code> and
<code>clasif</code>. This will be explained later.</p>
<p>From now on, typical calls to obtain only the initial medoids would
be</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>Lbuild<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;BUILD&quot;</span>,<span class="at">max_iter=</span><span class="dv">0</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/Rtmpao5EW9/TrapnellPearson.bin</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 255993892 Kib.</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting BUILD initialization method, serial version</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 50. TD=11914.134766</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 28. 8 reassigned points. TD=10251.591797</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 284. 79 reassigned points. TD=9409.731445</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 107. 0 reassigned points. TD=8996.875977</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 268. 21 reassigned points. TD=8661.666016</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 118. 0 reassigned points. TD=8333.292969</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 17. 108 reassigned points. TD=8016.939453</span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 154. 0 reassigned points. TD=7747.802734</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 126. 0 reassigned points. TD=7500.490234</span></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 173. 5 reassigned points. TD=7287.086914</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 7287.086914</span></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; BUILD initialization method (serial version) finished. Elapsed time: 0.00187154 s</span></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.00187154 s (method BUILD).</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0 s in 0 iterations.</span></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00187154 s (0 minutes, 0.00187154 seconds).</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>Llab1<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;LAB&quot;</span>,<span class="at">max_iter=</span><span class="dv">0</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/Rtmpao5EW9/TrapnellPearson.bin</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 255993640 Kib.</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting LAB initialization method, serial version.</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 50. TD=11914.134766</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 137. 48 reassigned points. TD=11207.690430</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 39. 111 reassigned points. TD=10894.427734</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 34. 9 reassigned points. TD=9396.723633</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 200. 55 reassigned points. TD=9186.223633</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 70. 5 reassigned points. TD=8898.829102</span></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 91. 21 reassigned points. TD=8642.791992</span></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 224. 7 reassigned points. TD=8542.523438</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 58. 8 reassigned points. TD=8445.713867</span></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 198. 37 reassigned points. TD=8297.508789</span></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 8297.508789</span></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; LAB initialization method (serial version) finished. Elapsed time: 0.000214903 s</span></span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.000214903 s (method LAB).</span></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0 s in 0 iterations.</span></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.000214903 s (0 minutes, 0.000214903 seconds).</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>Llab2<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;LAB&quot;</span>,<span class="at">max_iter=</span><span class="dv">0</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/Rtmpao5EW9/TrapnellPearson.bin</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 255993640 Kib.</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting LAB initialization method, serial version.</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 13. TD=12031.589844</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 143. 50 reassigned points. TD=11301.868164</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 22. 9 reassigned points. TD=9656.576172</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 88. 114 reassigned points. TD=9350.387695</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 108. 22 reassigned points. TD=9022.187500</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 173. 7 reassigned points. TD=8891.097656</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 178. 58 reassigned points. TD=8640.534180</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 109. 45 reassigned points. TD=8462.361328</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 209. 36 reassigned points. TD=8330.823242</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 26. 49 reassigned points. TD=8223.282227</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 8223.282227</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; LAB initialization method (serial version) finished. Elapsed time: 0.000217719 s</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.000217719 s (method LAB).</span></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0 s in 0 iterations.</span></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.000217719 s (0 minutes, 0.000217719 seconds).</span></span></code></pre></div>
<p>WARNING: the normal way of calling <code>ApplyPAM</code> would use
nthreads=0 to make use of all available cores in your machine.
Nevertheless, this does not seem to be allowed by CRAN to pass the test
so I have had to use the serial version invoked with nthreads=-1. In
your normal use of code try always nthreads=0. For the LAB method this
does not matter, since parallel implementation is not yet provided.</p>
<p>As you can see, to get and compare different initializations you must
set the parameter <code>max_iter</code> to value <code>0</code>. In this
case no iterations of objective function reduction are performed, and
the returned object contains the initial medoids and the classification
induced by them. Notice that even looking equal, the results of the
latter two calls are different since <code>LAB</code> initializes with a
random component (the sample to choose initial medoids is chosen
randomly).</p>
<p>You can check that the medoids, stored in <code>Llab1$med</code> and
<code>Llab2$med</code> (see more on this below) are in general,
different.</p>
<p>Now, these results can be used to initialize PAM if you find that any
of them contains a specially good set of medoids. This is the role of
method <code>PREV</code> that we have mentioned before. A typical call
would be</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>Llab2Final<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;PREV&quot;</span>,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">initial_med=</span>Llab2<span class="sc">$</span>med,<span class="at">nthreads=</span><span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/Rtmpao5EW9/TrapnellPearson.bin</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 255993892 Kib.</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting improved FastPAM1 method in serial implementation...</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 0. Medoid at place 4 (point 88) swapped with point 107; TD-change=-330.062347; TD=7893.220703. 43 reassigned points.</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 1. Medoid at place 2 (point 143) swapped with point 118; TD-change=-237.802246; TD=7655.418457. 16 reassigned points.</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 2. Medoid at place 8 (point 109) swapped with point 154; TD-change=-176.831970; TD=7478.586914. 28 reassigned points.</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 3. Medoid at place 1 (point 14) swapped with point 126; TD-change=-126.650116; TD=7351.936523. 43 reassigned points.</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 4. Medoid at place 7 (point 178) swapped with point 284; TD-change=-104.315048; TD=7247.621582. 6 reassigned points.</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 5. Medoid at place 10 (point 26) swapped with point 39; TD-change=-76.960526; TD=7170.661133. 14 reassigned points.</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 6. Medoid at place 3 (point 22) swapped with point 28; TD-change=-30.032187; TD=7140.628906. 0 reassigned points.</span></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Optimization method (serial version) finished. Elapsed time: 0.00192886 s</span></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0 s (method PREV).</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0.00192886 s in 6 iterations (0.000321476 seconds/iteration).</span></span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00192886 s (0 minutes, 0.00192886 seconds).</span></span></code></pre></div>
<p>where the initial set of medoids is taken from the object returned by
the former calls.</p>
<p>With respect to that object, as we said it is a list with two
vectors. The first one, <code>L$med</code>, has as many components as
requested medoids and the second, <code>L$clasif</code>, has as many
components as instances (here, cells).</p>
<p>Medoids are expressed in <code>L$med</code> by its number in the
array of points (row in the dissimilarity matrix) starting at 1 (R
convention).</p>
<p><code>L$clasif</code> contains the number of the medoid (i.e.: the
cluster) to which each instance has been assigned, according to their
order in <code>L$med</code> (also from 1).</p>
<p>This means that if <code>L$clasif[p]</code> is <code>m</code>, the
point <code>p</code> belongs to the class grouped around medoid
<code>L$med[m]</code>. Let us see it:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Which are the indexes of the points chosen as medoids?</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>L<span class="sc">$</span>med</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1269168 GSM1268987 GSM1269243 GSM1269066 GSM1269227 GSM1269077 GSM1268998 </span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        209         28        284        107        268        118         39 </span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1269113 GSM1269085 GSM1269132 </span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        154        126        173</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># In which class has point 147 been classified?</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>L<span class="sc">$</span>clasif[<span class="dv">147</span>]</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1269106 </span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;          7</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># And which is the index (row in the dissimilarity matrix) of the medoid closest to point 147?</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>L<span class="sc">$</span>med[L<span class="sc">$</span>clasif[<span class="dv">147</span>]]</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1268998 </span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         39</span></span></code></pre></div>
<p>In this way, values in L$clasif are between 1 and the number of
medoids, as we can see.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(L<span class="sc">$</span>clasif)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 1</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(L<span class="sc">$</span>clasif)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 10</span></span></code></pre></div>
<p>Therefore, they can be used as factors.</p>
<p>To help in the data analysis a data frame can be built, too, with the
names of the cells instead of their numbers. It is created as</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>F <span class="ot">&lt;-</span> <span class="fu">ClassifAsDataFrame</span>(L,tmppeafile1)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span></code></pre></div>
<p>The second parameter is the dissimilarity matrix created before when
we called <code>CalcAndWriteDissimilarityMatrix</code>. The reason of
using it here is because the returned dataframe has three columns:
<code>CellName</code> with name of each cell, <code>NNCellName</code>
with the name of the cell which is the center of the cluster to which
<code>CellName</code> belongs to and <code>NNDistance</code> which is
the distance between the cells <code>CellName</code> and
<code>NNCellName</code>.</p>
<p>This data frame implicitly knows which cells are medoids: those whose
distance to the closest medoid (themselves) is 0. They can be extracted
as</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract column 1 (cell name) of those rows for which distance to</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># closest medoid (column 3) is 0</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>F[<span class="fu">which</span>(F[,<span class="dv">3</span>]<span class="sc">==</span><span class="dv">0</span>),<span class="dv">1</span>]</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;GSM1268987&quot; &quot;GSM1268998&quot; &quot;GSM1269066&quot; &quot;GSM1269077&quot; &quot;GSM1269085&quot;</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [6] &quot;GSM1269113&quot; &quot;GSM1269132&quot; &quot;GSM1269168&quot; &quot;GSM1269227&quot; &quot;GSM1269243&quot;</span></span></code></pre></div>
<p>Finally, an abundance matrix can be obtained if the cells came from
different groups (samples) by combining such groups with the obtained
clusters. This is done by function
<code>BuildAbundanceMatrix</code>,</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>M<span class="ot">=</span><span class="fu">BuildAbundanceMatrix</span>(L<span class="sc">$</span>clasif,Gr)</span></code></pre></div>
<p>where the <code>Gr</code> vector should have been obtained before
(either from the Seurat object with <code>GetSeuratGroups</code>, or by
any other means from your input object, if possible). Since the data of
this example does not come from Seurat, this is not calculated now, but
in a real case the resulting object is a matrix with as many rows as
clusters and as many columns as groups. Each entry has the number of
cells of its sample which have been classified at that group. This
should be analyzed to determine if being in a different group has
sufficient influence on the cluster to which cells belong and therefore
if these clusters have any real biological significance.</p>
</div>
<div id="alternative-workflow-filtering-by-silhouette" class="section level2">
<h2>Alternative workflow: filtering by silhouette</h2>
<p>It is interesting to filter cells based on the degree in which they
belongs to a cluster. Indeed, cluster refinement can be done getting rid
of cells far away from any cluster center, or which are at a similar
distance of two or more of them.</p>
<p>This is characterized by the silhouette of each point (cell). Three
functions deal with this: <code>CalculateSilhouette</code>,
<code>FilterBySilhouetteQuantile</code> and
<code>FilterBySilhouetteThreshold</code>.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>S<span class="ot">=</span><span class="fu">CalculateSilhouette</span>(Llab2<span class="sc">$</span>clasif,tmppeafile1,<span class="at">nthreads=</span><span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 255997232 Kib.</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Calculating silhouette (serial implementation)...</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 288 points classified in 10 classes.</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finished serial implementation of silhouette (including dissimilarity matrix load). Elapsed time: 0.000262823 s</span></span></code></pre></div>
<p>WARNING: the normal way of calling <code>CalculateSilhouette</code>
would use nthreads=0 to make use of all available cores in your machine.
Nevertheless, this does not seem to be allowed by CRAN to pass the test
so I have had to use the serial version invoked with nthreads=-1. In
your normal use of code try always nthreads=0.</p>
<p>The parameters to function <code>CalculateSilhouette</code> are the
array of class membership, as returned by <code>ApplyPAM</code> in its
<code>clasif</code> field, and the file with the matrix of
dissimilarities.</p>
<p>A parallel implementation has been programmed, being nthreads as
explained before.</p>
<p>Silhouette is a number in <span class="math inline">\([-1,1]\)</span>; the higher its value, the most
centered a point is in its cluster.</p>
<p>The returned object <code>S</code> is a numeric vector with the value
of the silhouette for each point, which will be named if the
classification vector was named.</p>
<p>This vector can be converted to object of the class
<code>cluster:silhouette</code> with the function
<code>NumSilToClusterSil</code> (which needs the vector of
classifications, too). This is done so that, if you load the package
cluster, plot will generate the kind of silhouette plots included in
such package.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>Sclus <span class="ot">&lt;-</span> <span class="fu">NumSilToClusterSil</span>(Llab2<span class="sc">$</span>clasif,S)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Sclus)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAABlBMVEUAAAD///+l2Z/dAAAACXBIWXMAAA7DAAAOwwHHb6hkAAALIElEQVR4nO2di5akKgxFw///9KypUsgLDioqVOfcO92KIcIuCAjWDKVQU/R2AWZXAAIKQEABCCgAAQUgoAAEFICAAhBQAAIKQEABCCgAAQUgoAAEFICAAhBQAAIKQEABCCgAAQUgoAAEFICAAhBQAAIKQEABCCgAAQUgoAAENB4QfbUd8h833Gm8U3OT4Q6pEDoJiGXusGslDNBojwUNBSDX31bGUtYAJP3xQsoWVC6RxMjO9tbH+ql1KUIcMyhZRlZosD9RSgGIRyYRpJJI1oAYqZSMI+FuDUDVUYwEDxeQc033UOPIyTK2OsM9sk/SrbOtuEz24CUZ0ZzwrzgOrMxwj1+3qgZdgFhXSeyn6DnLA9rLGIBq/i4Akj78sPILgCj/8uOu13ZSzfj3gjTvE1VAzMAcb0eJn0hAwr1xN7o+g/3lAm+H7idc6rGjkMkFjbTeXAj31t3g6gz2d7fu6EXtGz57u8sKQEABCCgAzaYABBSAgAIQUAACCkBAAQhoHkDlcQraXLjJ4XnUbIAotcrkXumvwvaMe6RU8wCi+wGR+n0o09uixD5dp1DbeoloaHxtLq+vJb5Sou9xqmCzqAmo0MkWxNLLAZF0pW9wrljPiS9t6YNSnErdqA7Ipq4JyHzgsnYAkGpBPKQoQBeD+fVcJyWrISPKYUAyruf17M2wHYIO1Ho0IGpNNQYDYhGHtKWHaIog3THVEJVSXawyzA+JQQzzqy0Ied0s/CBdAaRHMdKALHXXy35hFkBqT0sZuC2o9qhBpTd9dn72g++1ki/j9ytGxc+hCo0TaEFE0go3uFs0LSDSVn8ckO5iZKz+OKD6zZwY9KSmBaS27cXBgzpyx+dLt5gCEFAAAgpAQAEIKAABzQEIlYLwIgGfm7Zci8sdk40lAMlFjaYLY0L2lMxJw/kcgNpSDya+zQlAcmFhakDkHH1O8tqGenITC6y0Wers1uHH1K5Atab/cwPikYecGnFiqvauw0TZmBusBMhJlj1BXi79g/+su7aASo61ATmjGDmArDcDSMYy5ndhQES6CntsyoC2dQJe3+11feORRy/dRruL9rAQIG3Gex1Jm8RbhRn5BBdnOOsu2sNqA1Jf9UlVQHxs2xnouPVpZsI8px4p2sOqDvOfn2qTOSMgdpEoBxjRHnQX+3wXhoeizHdpQFuAEVyIvcmRctNQjczGIMqxia2Yw93gtwVKISqrApG0M4N2Y1NOfCrVEqwESM1yOgBRGxBrbasDcr5ap0y2+MQeJJxhnrKP3Mty6rmiPSK4r8EnOLXBOMcnhYSsGamTMq65nueXLL5b4gyIZK8yi0McEGWLxQHtMULO8hwDNvznK0574u885LGs+rrDpZI/pF5AjoH3qCHGuX0ZoN5551eue20UI2HUiEFyqkn79Z8BpBaUrYHtKXYmzWjkffC1AZXo0mFgn72sKZtrkpfq5JhazRiqDIxRExC5qcrBscK+ILbQUzVI/mzPzIoSmz7zLGTzMh+Ty3y8lsI+23Pai9fliGfJoXrZYR4A2i/np3LWFrxOtzdH4tZ8j8S9/cTycNgUckC6gIRVK2/9jpNJhudacck16AbktrX2HadRvejKSB7U8lYArRyk/wuUssLnb7Sgj5qlZGv6rWxqs7r8WjwGfQTn0b6Rk+3nALUmufJSlYZNM/OAhQGp6NBoJUcBkXfS52Musc0974mDlMQl350zeNWf9eYHtCl/0Ngqn1QaFW+Q5KTWPU4sGIuc692AWs7XAlQ7rRjVIhaVsx8C1PHIcRbQAZ8zqoTf1mi/XeYYnbBF/EUOsTa0epD23q7Tl3dGJdFk2RY3ykke0mqulwHUHsUoI2JWIgvbieZQ2AtCNc9riM9d3KsIkHGjlrL/BqCDXSyxZrQ4IB2CbLtw+koTEHFAxk76XkCUWsN8AQRbkFiTTrkFLb43D0pJ8kdvDMonzfj/E4DUROkwoF9vQcUEArJ9UBifu/X7wjsbfNmiFYOKwR6HfmIe1BrB9iT2FiOpS+KUAxJvcVZIrABIdpyWgWPZ6mKlAdX9LwjIFlkBcq5VXTmhq5V/UsEutqc7b2k0AbmxvZV/VrXeDso2qcyRVbI++zlArUcBboNj0G8CgkFaBNsAVLUIQG0D37I9zLMcCw/zcBSjiqVjnQ2IX14dENrRYA+bYJgXk2f+a+1HjSu6WsEAdHP++wUnidnArOm4oUV2V0J9bHpAcJIow64ZwUwM4mki75rDfOcQTz1DfE4zgBp3WQxQbYgntybVJD3G/zKgnAyf5K2/HueLAXJN2Bctm5NE4086XzMGfRsHKmUljhwCtOoolswqs2uEAbXek258BvMD+q9xo1gFUP0GswN6cphv+J9XavZXaROVyV4TEKX1AX0DtNkL1FbsSV59J9wFWt4JygGusfc8PSD4EacCQlv60d1aNYynBoR39XZD3hPJpBtrabQsIL/afhQ6AMj4U2HeNZ9UTwAy46BrPqn8antWNctuQNV+/EuAtlHPeHCc/iVAfKbUfvnFewstmRbouZ9VfYCoMVXqbUH1t/DmB9QxzB+KQbV50LItSFXILzDJH01r3aWWB9T/RSjbSfw6u3uHKwK6ruvVC0C3e5haAQgoAN2uAAQUgIACEFAAAgpAQAEIKAABBSCgAAQUgIACEFAAAgpAQAEIKAABBSCgAAQUgIACENDtgGhSdZf/Tjgnb3A8y405AtBw1wcVgMbfIAANzxKA3ssRgIa7PqgANP4GfwvQ6gpAQAEIKAABBSCgAAQUgIACEFAAAgpAQAEIKAAB3QmI1F+n1v23tZmTUXf4mB27x42Aer4IV7WXmQfdQZl13+MmzQmIpgFEwn9vWyYv86A7fK3EZ4CzPQYIRyBufwZQ53bprIDgvUa0oJ7aTAoI3+wqoJ4cwuQ9QN/GHoC6PAegtms5xOAgdH6Y76uuMJlmHkSpqzBH7c/kSJzk24Dk33DQMQgftT+TQzS1lx81fkMBCCgAAQUgoAAEFICAAhBQAAIKQEABCCgAAQUgoAAEFICAngGUNz1p/3+MW31wQ20eASTWqHoB5ZUw5Le4XhWQWRUdBUjY/gag/f+8y1e63v5zT/n/ezMj9Zchs96a+L/8wB1nH4MKf6vEOvmnMnuq/OjFmfwhWiEH9AXIDxzbIWW/U8TKTAxH40Cz4bVmf7wl+AUB8a50HVDel0SARlTuKUDJjmL+Qf7SdgEkvsf9DWFdgFjfu1bqu+UE6XLqHuRcTgVp+y91ABrQiuYE1OhiBwDRbj6i8LdK1MKt1bcu8kz8IOlNeVgdkP+osUGhbCFTtmoTT8r5lIedr25s7DWT00W/lHsJBSCgAAQUgG5VAAIKQEABCCgAAQUgoAAEFICAAhBQAAIKQEABCCgAAQUgoAAEFICAAhDQDYCGbGiK5f0h3uRBh2MasO1YcTzESz+gskHUNtkP+upddqkGa5zTgYCE7SSA7OYXfRpX2Wi3W2HbCd9ifOhNIq8a6fq2oycLyGyAVvdEk6gS7d2V1B91Jn+QcpRLJO8nm92DgGxpc5VIWJE5EPmpYuRkU2y4J/bHIduq/buAyK/pfqlcuAio902iWj1uAMT+VYbSmkn8XQx7KGC1UG8W7hfagIa+SaSqcCMg9qu8pCISc+HFgSxidwsSfpWnY28SuRV5EBAl78zpYiX1Yhe7Bkg2vGESrYV/7V9URA0mpDM6gJQTx2XSnjQJ537yrrYqNwMqrUJMYb7Tli2Z2PVUrltAN79J5NXlrkeNQ3r7/lDvFRBHgSn0MqDp+bxZwhHfpLhfCxTxXQUgoH/FKDRMTI+1nwAAAABJRU5ErkJggg==" /><!-- --></p>
<p>Once the silhouette is calculated we can filter it by quantile or by
threshold. All points under this quantile or threshold will be
discarded, except if they are medoids. Parameters are the silhouette, as
returned by <code>CalculateSilhouette</code>, the list of
medoids/clasif, as returned by <code>ApplyPAM</code>, the file with
matrix of counts for the whole set of cells, the file that will contain
the matrix of counts of the remaining cells, the file with dissimilarity
matrix for the whole set, the file that will contain the dissimilarity
for the remaining cells and (depending on the function used) the
quantile in <span class="math inline">\([0,1]\)</span> or the silhouette
threshold in <span class="math inline">\([-1,1]\)</span>. As an
example:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>tmpfiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellFilt.bin&quot;</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>tmppeafiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearsonFilt.bin&quot;</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>Lfilt<span class="ot">=</span><span class="fu">FilterBySilhouetteQuantile</span>(S,Llab2,tmpfile1,tmpfiltfile1,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                                 tmppeafile1,tmppeafiltfile1,<span class="fl">0.2</span>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; After filtering silhouette with quantile 0.2 (threshold -0.163434) 232 of the 288 points remain.</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in FilterBySilhouetteQuantile(S, Llab2, tmpfile1, tmpfiltfile1, : One</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; or more medoids have been kept, even they were below the threshold (which seems</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; problematic. Check your clusters...).</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 of the medoids have been kept, even they were below the threshold (which seems problematic. Check your clusters...).</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/Rtmpao5EW9/TrapnellFilt.bin of (232x5218)</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 2385968</span></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (232 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing column names (5218 strings written, from ENSG00000164675.10 to ENSG00000093100.13).</span></span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Experiment conquer GSE52529-GPL16791 Filtered by silhouette from file /tmp/Rtmpao5EW9/Trapnell.bin with quantile 0.2. </span></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/Rtmpao5EW9/TrapnellPearsonFilt.bin</span></span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 108240</span></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (232 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Pearson dissimilarities for cell</span></span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          in experiment conquer</span></span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          GSE52529-GPL16791 Filtered by silhouette from file /tmp/Rtmpao5EW9/TrapnellPearson.bin with quantile 0.2.</span></span></code></pre></div>
<p>If the original matrix contained row (in this case, cell) and column
(in this case, gene) names, the column names are copied and the row
names are transported for those rows (cells) that remain. The same
happens with respect to rows of the dissimilarity matrix.</p>
<p>Notice that the new dissimilarity matrix could have been calculated
from the matrix of filtered counts with
<code>CalcAndWriteDissimilarityMatrix</code> but creating it here,
simply getting rid of the filtered rows and columns is much faster.</p>
<p>Also, if a medoid is below the silhouette quantile, it will not be
filtered out, but a warning message will be shown, since this is a
strange situation that may indicate that some of your clusters are not
real but artifacts due to a few outliers that are close to each
other.</p>
<p>Of course you can use the filtered cells to build a data frame, as
long as the dissimilarity matrix contains cell names,</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>tmppeafiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearsonFilt.bin&quot;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>Ffilt<span class="ot">=</span><span class="fu">ClassifAsDataFrame</span>(Lfilt,tmppeafiltfile1)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (232,232)</span></span></code></pre></div>
<p>or to recreate a .csv file of counts of the remaining cells with
their correct names using <code>BinMatToCsv</code> like this.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>tmpfiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellFilt.bin&quot;</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>tmpfiltcsvfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellFilt.csv&quot;</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatToCsv</span>(tmpfiltfile1,tmpfiltcsvfile1)</span></code></pre></div>
<p>But remember that this was the result of the first step of the PAM
algorithm, so probably you will want to make them iterate.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>tmppeafiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearsonFilt.bin&quot;</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>Lfinal<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafiltfile1,<span class="at">k=</span><span class="fu">length</span>(Lfilt<span class="sc">$</span>med),<span class="at">init_method=</span><span class="st">&quot;PREV&quot;</span>,<span class="at">initial_med=</span>Lfilt<span class="sc">$</span>med,<span class="at">nthreads=</span><span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/Rtmpao5EW9/TrapnellPearsonFilt.bin</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 105 KiB, which is 0% of the available memory, which is 256015172 Kib.</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (232,232)</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 232 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting improved FastPAM1 method in serial implementation...</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 0. Medoid at place 1 (point 10) swapped with point 76; TD-change=-467.832794; TD=8434.242188. 5 reassigned points.</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 1. Medoid at place 4 (point 61) swapped with point 87; TD-change=-317.871826; TD=8116.370605. 44 reassigned points.</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 2. Medoid at place 6 (point 132) swapped with point 117; TD-change=-241.221786; TD=7875.148926. 3 reassigned points.</span></span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 3. Medoid at place 8 (point 78) swapped with point 94; TD-change=-172.342545; TD=7702.806152. 28 reassigned points.</span></span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 4. Medoid at place 7 (point 136) swapped with point 126; TD-change=-110.021385; TD=7592.784668. 14 reassigned points.</span></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 5. Medoid at place 2 (point 109) swapped with point 158; TD-change=-65.576607; TD=7527.208008. 13 reassigned points.</span></span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 6. Medoid at place 3 (point 16) swapped with point 20; TD-change=-37.281467; TD=7489.926758. 0 reassigned points.</span></span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Optimization method (serial version) finished. Elapsed time: 0.00115008 s</span></span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0 s (method PREV).</span></span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0.00115008 s in 6 iterations (0.00019168 seconds/iteration).</span></span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00115008 s (0 minutes, 0.00115008 seconds).</span></span></code></pre></div>
<p>WARNING: the normal way of calling <code>ApplyPAM</code> would use
nthreads=0 to make use of all available cores in your machine.
Nevertheless, this does not seem to be allowed by CRAN to pass the test
so I have had to use the serial version invoked with nthreads=-1. In
your normal use of code try always nthreads=0.</p>
<p>Of course, we might have used simply 10 as number of medoids,
<code>k</code>, since this does not change by filtering, but this is to
emphasize the fact that <code>ApplyPAM</code> with method
<code>PREV</code> requires both parameters to be consistent.</p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-R-Matrix" class="csl-entry">
Bates, Douglas, Martin Maechler, and Mikael Jagan. 2022. <em>Matrix:
Sparse and Dense Matrix Classes and Methods</em>. <a href="https://CRAN.R-project.org/package=Matrix">https://CRAN.R-project.org/package=Matrix</a>.
</div>
<div id="ref-R-jmatrix" class="csl-entry">
Domingo, Juan. 2023a. <em>Jmatrix: Read from/Write to Disk Matrices with
Any Data Type in a Binary Format</em>.
</div>
<div id="ref-R-parallelpam" class="csl-entry">
———. 2023b. <em>Parallelpam: Applies the Partitioning-Around-Medoids
(PAM) Clustering Algorithm to Big Sets of Data Using Parallel
Implementation, If Several Cores Are Available</em>.
</div>
<div id="ref-Domingo2023" class="csl-entry">
Domingo, Juan, Teresa Leon, and Esther Dura. 2023. <span>“Scellpam: An
<span>R</span> Package/<span>C</span>++ Library to Perform Parallel
Partitioning Around Medoids on scRNAseq Data Sets.”</span> <em>BMC
Bioinformatics</em> 24 (1): 342. <a href="https://doi.org/10.1186/s12859-023-05471-1">https://doi.org/10.1186/s12859-023-05471-1</a>.
</div>
<div id="ref-R-DuoClustering2018" class="csl-entry">
Duò, Angelo, and Charlotte Soneson. 2021. <em>DuoClustering2018: Data,
Clustering Results and Visualization Functions from Duò Et Al
(2018)</em>.
</div>
<div id="ref-R-Seurat" class="csl-entry">
Hao, Yuhan, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III,
Shiwei Zheng, Andrew Butler, Maddie J. Lee, et al. 2021.
<span>“Integrated Analysis of Multimodal Single-Cell Data.”</span>
<em>Cell</em>. <a href="https://doi.org/10.1016/j.cell.2021.04.048">https://doi.org/10.1016/j.cell.2021.04.048</a>.
</div>
<div id="ref-Schubert2019" class="csl-entry">
Schubert, Erich, and Peter J. Rousseeuw. 2019. <span>“Faster k-Medoids
Clustering: Improving the <span>PAM</span>, <span>CLARA</span>, and
<span>CLARANS</span> Algorithms.”</span> In <em>Similarity Search and
Applications</em>, 171–87. Springer International Publishing. <a href="https://doi.org/10.1007/978-3-030-32047-8_16">https://doi.org/10.1007/978-3-030-32047-8_16</a>.
</div>
<div id="ref-R-splatter" class="csl-entry">
Zappia, Luke, Belinda Phipson, and Alicia Oshlack. 2017.
<span>“Splatter: Simulation of Single-Cell RNA Sequencing Data.”</span>
<em>Genome Biology</em>. <a href="https://doi.org/10.1186/s13059-017-1305-0">https://doi.org/10.1186/s13059-017-1305-0</a>.
</div>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
