<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Juan Domingo" />


<title>scellpam</title>

<script src="data:application/javascript;base64,Ly8gUGFuZG9jIDIuOSBhZGRzIGF0dHJpYnV0ZXMgb24gYm90aCBoZWFkZXIgYW5kIGRpdi4gV2UgcmVtb3ZlIHRoZSBmb3JtZXIgKHRvCi8vIGJlIGNvbXBhdGlibGUgd2l0aCB0aGUgYmVoYXZpb3Igb2YgUGFuZG9jIDwgMi44KS4KZG9jdW1lbnQuYWRkRXZlbnRMaXN0ZW5lcignRE9NQ29udGVudExvYWRlZCcsIGZ1bmN0aW9uKGUpIHsKICB2YXIgaHMgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yQWxsKCJkaXYuc2VjdGlvbltjbGFzcyo9J2xldmVsJ10gPiA6Zmlyc3QtY2hpbGQiKTsKICB2YXIgaSwgaCwgYTsKICBmb3IgKGkgPSAwOyBpIDwgaHMubGVuZ3RoOyBpKyspIHsKICAgIGggPSBoc1tpXTsKICAgIGlmICghL15oWzEtNl0kL2kudGVzdChoLnRhZ05hbWUpKSBjb250aW51ZTsgIC8vIGl0IHNob3VsZCBiZSBhIGhlYWRlciBoMS1oNgogICAgYSA9IGguYXR0cmlidXRlczsKICAgIHdoaWxlIChhLmxlbmd0aCA+IDApIGgucmVtb3ZlQXR0cmlidXRlKGFbMF0ubmFtZSk7CiAgfQp9KTsK"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="data:text/css,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">scellpam</h1>
<h4 class="author">Juan Domingo</h4>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scellpam)</span></code></pre></div>
<div id="purpose" class="section level1">
<h1>Purpose</h1>
<p>The package <code>scellpam</code> (<span class="citation">Domingo (2022c)</span>) is meant as a way to apply the PAM algorithm to results of single-cell RNA-Seq data sets.</p>
<p>Differently to other packages, it can deal with a big number of cells (limited in principle only by the amount of available RAM) and make the calculations in parallel.</p>
<p>It uses a special format to store binary matrices on disk, the <code>jmatrix</code> format. Please, make yourself familiar with the <code>jmatrix</code> creation and manipulation and with the PAM application looking first at the other vignettes of this package, <code>jmatrixsc</code> and <code>parallelpamsc</code>.</p>
<p>WARNING: you must NOT load explicitly neither <code>jmatrix</code> nor <code>parallelpam</code>. Indeed, you do not need even to install them. All their functions have been included here, too, so doing <code>library(scellpam)</code> is enough.</p>
</div>
<div id="workflow" class="section level1">
<h1>Workflow</h1>
<div id="debug-messages" class="section level2">
<h2>Debug messages</h2>
<p>First of all, the package can show quite informative (but sometimes verbose) messages in the console. To turn on/off such messages you can use,</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Initially, state of debug is FALSE. Turn it on with</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ScellpamSetDebug</span>(<span class="cn">TRUE</span>,<span class="at">debparpam=</span><span class="cn">TRUE</span>,<span class="at">debjmat=</span><span class="cn">TRUE</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Debugging for scellpam (biological part) of the package set to ON.</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Debugging for parallelpam inside scellpam package set to ON.</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Debugging for jmatrix inside scellpam package set to ON.</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># We have also turned on debugging of the parallel PAM algorithm and of the binary matrix creation/manipulation</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># but if this annoys you their default value is FALSE for both cases even when the scellpam debug is set to true, so just do</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ScellpamSetDebug(TRUE)</span></span></code></pre></div>
</div>
<div id="data-loadstorage" class="section level2">
<h2>Data load/storage</h2>
<p>As stated before, the limitations in memory suggested us to use intermediate binary files stored in disk as a way to keep the results of each step of the process. These files have a concrete internal format; see the vignette <code>jmatrixsc</code> in this package to know more about the <code>jmatrix</code> (<span class="citation">Domingo (2022a)</span>) package.</p>
<p>Thus, the first step is to load raw data (sequencing counts) coming either from a .csv file or from the main formats used in single cell RNA-Seq data sets and store them as a binary matrix file in <code>jmatrix</code> format.</p>
<p>The input formats can be a <code>.csv</code> file, a sparse matrix stored into the S4 object created by the Seurat package (<span class="citation">Hao et al. (2021)</span>) or a R NumericMatrix extracted from any package that uses a single-cell experiment object. Unfortunately, the internal structure of such objects does not seem to be the same for all packages so we have not been able to provide a single function to extract from all of them.</p>
<div id="readwrite-.csv-files" class="section level3">
<h3>Read/write .csv files</h3>
<p>Assuming a <code>.csv</code> input file, the following call reads the file and writes to disk its binary representation. The <code>jmatrix</code> format can create full, sparse or symmetric matrices.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToJMat</span>(<span class="st">&quot;countsfile.csv&quot;</span>,<span class="st">&quot;countsfile.bin&quot;</span>)</span></code></pre></div>
<p>By default, data are stored as a sparse matrix of numbers as floats (4 bytes per item in most architectures). Indeed, default parameters are</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToJMat</span>(<span class="st">&quot;countsfile.csv&quot;</span>,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">mtype=</span><span class="st">&quot;sparse&quot;</span>,<span class="at">csep=</span><span class="st">&quot;,&quot;</span>,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">ctype=</span><span class="st">&quot;raw&quot;</span>,<span class="at">valuetype=</span><span class="st">&quot;float&quot;</span>,<span class="at">transpose=</span><span class="cn">FALSE</span>,<span class="at">comment=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>where mtype can be <code>sparse</code> or <code>full</code> and valuetype can be <code>float</code> or <code>double</code>. Changing the default values for these parameters is strongly discouraged, except when the sparse matrix is bigger than the full matrix, which happens when there are few zeros. In such a case use <code>mtype=&quot;full&quot;</code>. The names for rows and columns are read from the <code>.csv</code> file and stored as metadata in the binary file after the raw data of counts.</p>
<p>Other parameters you may (and need to) change are the separation character for fields in the <code>.csv</code> file (csep) if your fields are not comma-separated, the type of normalization (ctype) which can be <code>raw</code> to write the raw value of the counts, <code>log1</code> to write the <code>log2(counts+1)</code>, <code>rawn</code> which is as <code>row</code> but normalized and <code>log1n</code> which is like <code>log1</code> but normalized.</p>
<p>WARNING: normalize ALWAYS DOES IY BY COLUMNS (before transposition, if requested to transpose). The logarithm is taken base 2.</p>
<p>Finally, you may want to write the matrix transposed and add any comment of up to 1024 characters as a string, if you want. Such comment will be stored in the metadata, after the row and column names.</p>
<p>In our case, the data are organized with genes by rows and cells by columns (that is why we choose to normalize by colums), which is not what is needed later to calculate the dissimilarity matrix. Therefore, our typical call would be</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToBinMat</span>(<span class="st">&quot;countsfile.csv&quot;</span>,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">comment=</span><span class="st">&quot;Obtained from countsfile.csv&quot;</span>)</span></code></pre></div>
<p>As a result of the calculations the functions of this package generate binary files, also in <code>jmatrix</code> format. If you want to inspect them they can be written back as <code>.csv</code> files, which allows them to be imported as R tables, if your computer has memory enough to hold them. This is done as</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatToCsv</span>(<span class="st">&quot;countsfile.bin&quot;</span>,<span class="st">&quot;countsback.csv&quot;</span>,<span class="at">csep=</span><span class="st">&quot;,&quot;</span>,<span class="at">withquotes=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>Also, you can selectively read some rows and/or columns of them. See appropriate functions in the vignette called <code>jmatrixsc</code> attached to this package.</p>
<p>Notice that in this case <code>counstfile.csv</code> and <code>countsback.csv</code> would not be equal; row and column names will have been swapped and that is OK, since we wrote the binary file with <code>transpose=TRUE</code>.</p>
<p>Also, remember that if you use comma as the separator character, you can just get rid of the csep parameter. The parameter withquotes, if you decide to use it and you set it to TRUE, writes each row and column name in the <code>.csv</code> surrounded by double quotes (but not the numbers).</p>
<p>It is fair to have <code>.csv</code> files since they can be inspected to check errors but obviously these files can be quite large and writing them may be quite slow.</p>
</div>
<div id="readwrite-seurat-files" class="section level3">
<h3>Read/write Seurat files</h3>
<p>If your data are not in a <code>.csv</code> file but stored in a S4 object created by package <code>seurat</code> (<span class="citation">Hao et al. (2021)</span>) (and frequently stored for example in a compressed <code>.rds</code> file) we provide a function, too, to write it to <code>jmatrix</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;yourdata.rds&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dgCMatToJMat</span>(q<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>counts,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">comment=</span><span class="st">&quot;Obtained from yourdata.rds&quot;</span>)</span></code></pre></div>
<p>Nevertheless, sometimes the matrix is stored in a different slot of the object like</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dgCMatToJMat</span>(q<span class="sc">@</span>raw.data,<span class="st">&quot;countsfile.bin&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">comment=</span><span class="st">&quot;Obtained from other data&quot;</span>)</span></code></pre></div>
<p>It is up to you to inspect the original object (with <code>str(q)</code>) and see what you must do.</p>
<p>Parameters <code>mtype</code>, <code>ctype</code> and <code>valuetype</code> are available, too, with the same meaning as for the former function. The <code>.bin</code> file can obviously be dumped to <code>.csv</code>, as explained before.</p>
<p>Also, in the case of Seurat, each cell belongs to a sample, being the sample a set of cells with some common characteristic like coming from the same living being or from a common stage in a process of cellular change. Cells are then labelled with a factor and this could be obtained with</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>Gr<span class="ot">=</span><span class="fu">GetSeuratGroups</span>(q)</span></code></pre></div>
<p>where <code>Gr</code> is a numeric vector of integers with as many components as cells that contains a consecutive numeric identifier of the group starting at 1. But again this may depend on the internal structure of that particular Seurat object.</p>
</div>
<div id="readwrite-from-other-single-cell-experiment-packages" class="section level3">
<h3>Read/write from other single cell experiment packages</h3>
<p>Similarly, if you are using packages that use the <code>sce</code> (Single-cell experiment) S4 object (like <code>splatter</code> (<span class="citation">Zappia, Phipson, and Oshlack (2017)</span>), <code>DuoClustering</code> (<span class="citation">Duò and Soneson (2021)</span>) and others) you will have to extract the matrix of counts. Unfortunately, again this does not seem to be uniform since each package has a slightly different internal structure for the S4 object. The most we have been able to do is to provide a function which takes the sparse matrix of counts (as a numeric matrix) and the name of the generated binary file (as before) and optionally the row and column names (as <code>StringVectors</code>) if they were not attached to the matrix itself, followed by the parameters used for the csv/Seurat formats.</p>
<p>The next examples are not marked for execution since they need the loading of packages splatter/scatter and DuoClustering2018 respectively and a test to see if they are available using installed.packages is forbidden if you want to pass the CRAN tests. It you want to run them, please copy+paste this code in R.</p>
<p>Your code, for example with <code>splatter</code>, would probably look like</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">library</span>(splatter)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">library</span>(scater)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>Splattersce <span class="ot">&lt;-</span> <span class="fu">mockSCE</span>()</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu">SceToJMat</span>(Splattersce<span class="sc">@</span>assays<span class="sc">@</span>data<span class="sc">@</span>listData<span class="sc">$</span>counts,<span class="st">&quot;mockSCE.bin&quot;</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">mtype=</span><span class="st">&quot;full&quot;</span>,<span class="at">ctype=</span><span class="st">&quot;log1n&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">comment=</span><span class="st">&quot;Generated by splatter with mockSCE() and normalized</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="st">                    to log2(counts+1).&quot;</span>)</span></code></pre></div>
<p>Nevertheless, in <code>DuoClustering</code> you would have something like</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">library</span>(DuoClustering2018)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>KuTCCsce <span class="ot">&lt;-</span> <span class="fu">sce_filteredExpr10_KumarTCC</span>()</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">SceToJMat</span>(KuTCCsce<span class="sc">@</span>assays<span class="sc">$</span>data<span class="sc">@</span>listData<span class="sc">$</span>counts,<span class="st">&quot;KuTCC.bin&quot;</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">mtype=</span><span class="st">&quot;full&quot;</span>,<span class="at">ctype=</span><span class="st">&quot;log1n&quot;</span>,<span class="at">transpose=</span><span class="cn">TRUE</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">comment=</span><span class="st">&quot;Generated by DuoCLustering with Kumar data and</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="st">                normalized to log2(counts+1).&quot;</span>)</span></code></pre></div>
<p>Please, notice the subtle difference between</p>
<p><code>Splattersce@assays@data@listData$counts</code> and</p>
<p><code>KuTCCsce@assays$data@listData$counts</code></p>
<p>We have noticed that in simulations with <code>splatter</code>, and depending on the parameters, the number of zeros can be not too high and therefore storing the binary matrix as sparse matrix makes its size bigger, instead of smaller. In such a case you can store it as a full matrix passing the parameter <code>mtype=&quot;full&quot;</code> to <code>SceToJMat</code>.</p>
<p>Parameters <code>ctype</code> and <code>valuetype</code> are available, too, with the same meaning as for the former functions.</p>
</div>
<div id="getting-information-on-the-stored-matrix" class="section level3">
<h3>Getting information on the stored matrix</h3>
<p>Other useful function in this step is <code>JMatInfo</code>, which prints in the console or in a text file information about a stored binary file. As an example, let us load and write a <code>.csv</code> file and then let us see the information of the generated binary file.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>tmpfile<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">CsvToJMat</span>(<span class="st">&quot;extdata/conquer_GSE52529_Trapnell_sample.csv&quot;</span>,tmpfile,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">transpose=</span><span class="cn">TRUE</span>,<span class="at">comment=</span><span class="st">&quot;Experiment conquer GSE52529-GPL16791&quot;</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 289 columns (excluding column of names) in file extdata/conquer_GSE52529_Trapnell_sample.csv.</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5218 lines (excluding header) in file extdata/conquer_GSE52529_Trapnell_sample.csv</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data will be read from each line and stored as float values.</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read 5218 data lines of file extdata/conquer_GSE52529_Trapnell_sample.csv, as expected.</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Transposing matrix of (5218x288) to a matrix of (288x5218)</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/RtmpAg94fG/Trapnell.bin of (288x5218)</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 3030456</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (288 strings written, from &quot;GSM1268960&quot; to &quot;GSM1269247&quot;).</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing column names (5218 strings written, from &quot;ENSG00000164675.10&quot; to &quot;ENSG00000093100.13&quot;).</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Experiment conquer GSE52529-GPL16791</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatInfo</span>(tmpfile)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; File:               /tmp/RtmpAg94fG/Trapnell.bin</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Matrix type:        SparseMatrix</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of elements: 1502784</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data type:          float</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Endianness:         little endian (same as this machine)</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of rows:     288</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of columns:  5218</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata:           Stored names of rows and columns.</span></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata comment:  &quot;Experiment conquer GSE52529-GPL16791&quot;</span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Binary data size:   3030328 bytes, which is 50.4119% of the full matrix size.</span></span></code></pre></div>
<p>If you want to keep the information about the binary file in a text file, just do</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>tmpfile<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>tmptxtfile<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellInfo.txt&quot;</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatInfo</span>(tmpfile,tmptxtfile)</span></code></pre></div>
</div>
<div id="filtering-genes-or-cells" class="section level3">
<h3>Filtering genes or cells</h3>
<p>The binary files can be expurged getting rid of some genes or cells. The function provided for this task is <code>FilterBinByName</code>, which takes as inputs a binary file and a R list of R-strings with the names of the genes or cells that must remain (first and second parameters). The third one is the name of the binary filtered file, which will have the same nature (full or sparse matrix) and data type of the matrix in the original file. The fourth parameter, <code>namesat</code>, indicates if whatever you want to filter (genes or cells) are in columns (“cols”) or in rows (“rows”). As an example, let us filter some 1000 of the 65218 genes contained in file conquer_<code>GSE52529-GPL16791_Trapnell.csv</code> and whose names we have put in the list stored in variable <code>v</code> by reading the file <code>Trapnell_remain_genes.csv</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>v<span class="ot">&lt;-</span><span class="fu">as.vector</span>(<span class="fu">read.table</span>(<span class="st">&quot;extdata/Trapnell_remain_genes.csv&quot;</span>)[[<span class="dv">1</span>]])</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>tmpfiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell_filtered.bin&quot;</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">FilterJMatByName</span>(tmpfile1,v,tmpfiltfile1,<span class="at">namesat=</span><span class="st">&quot;cols&quot;</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1000 columns of the 5218 in the original matrix will be kept.</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/RtmpAg94fG/Trapnell_filtered.bin of (288x1000)</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 1276144</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (288 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing column names (1000 strings written, from ENSG00000067113.16 to ENSG00000093100.13).</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Experiment conquer GSE52529-GPL16791</span></span></code></pre></div>
<p>If the list contains names not present in the original file the program emits a warning (and keeps only whose which are present).</p>
</div>
</div>
<div id="calculating-the-distancedissimilarity-matrix" class="section level2">
<h2>Calculating the distance/dissimilarity matrix</h2>
<p>This is the most computationally intensive part of the process (particularly, for samples with a high number of cells) and therefore has been programmed in parallel, taking advantage of the multiple cores of the machine, if available. See vignette <code>parallelpamsc</code> of this package to know more about <code>parallelpam</code> (<span class="citation">Domingo (2022b)</span>). From now on, just the prospective call to comment on the parameters.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">CalcAndWriteDissimilarityMatrix</span>(tmpfile1,tmppeafile1,<span class="at">distype=</span><span class="st">&quot;Pearson&quot;</span>,<span class="at">nthreads=</span><span class="dv">0</span>,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">comment=</span><span class="st">&quot;Pearson dissimilarities for cell in experiment conquer GSE52529-GPL16791&quot;</span>)</span></code></pre></div>
<p>The input and output files (first and second parameters) are of course compulsory. Input file can be a sparse of full binary matrix (but obviously, not a symmetric matrix).</p>
<p>WARNING: notice that the vectors to calculate dissimilarities amongst them MUST be stored BY ROWS. This is due to efficiency reasons and it is the reason by which we transposed the counts matrix before writing it as binary.</p>
<p>Output dissimilarity matrix will always be a binary symmetric (obviously square) matrix with a number of rows (and columns) equal to the number of rows (in this case, cells) of the counts file. The type of distance/dissimilarity can be <code>L1</code> (Manhattan distance), <code>L2</code> (Euclidean distance) or <code>Pearson</code> (Pearson dissimilarity coefficient). The resulting matrix stores only names for the rows, which are the names of the cells stored as rows in file <code>Trapnell.bin</code>. If the number of cells is <span class="math inline">\(N\)</span>, only <span class="math inline">\(N(N+1)/2\)</span> dissimilarity values are really stored.</p>
<p>A note on the number of threads, valid also for other algorithms that will be explained later. Possible values for the number of threads is:</p>
<ul>
<li><p><code>-1</code> (or any negative number) to indicate you do not want to use threads (strictly sequential computation).</p></li>
<li><p><code>0</code> to allow the program to choose the number of threads according to the problem size and the number of available cores.</p></li>
<li><p>Any positive number to force the use of such number of threads.</p></li>
</ul>
<p>Choosing explicitly a number of threads bigger than the number of available<br />
cores is allowed, but discouraged and the program emits a warning about it.</p>
<p>With respect to option <code>0</code> (the program chooses), for small problems (in this case, less than 1000 cells) the function makes the choice of not using threads, since the overhead of opening and waiting termination is not worth. For bigger problems the number of chosen threads is the number of available cores. Nevertheless, this choice may not be the best, depending on your machine.</p>
<p>Now, let us really do the call with this small dataset.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">CalcAndWriteDissimilarityMatrix</span>(tmpfile1,tmppeafile1,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">comment=</span><span class="st">&quot;Pearson dissimilarities for cell</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="st">                                         in experiment conquer</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="st">                                         GSE52529-GPL16791&quot;</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Input matrix is a sparse matrix  with elements of type &#39;float&#39; and size (288,5218)</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read sparse matrix from file /tmp/RtmpAg94fG/Trapnell.bin.</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Its size is [288 x 5218] and it uses 378647 elements, half of 4 bytes and half of 4 bytes each, with accounts for 2.88995 MBytes.</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Creating dissimilarity matrix of size (288x288)</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 250891632 Kib.</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of dissimilarity matrix calculation Elapsed time: 0.729654 s</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/RtmpAg94fG/TrapnellPearson.bin</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 166592</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (288 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Pearson dissimilarities for cell</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          in experiment conquer</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          GSE52529-GPL16791</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Output binary file /tmp/RtmpAg94fG/TrapnellPearson.bin written.</span></span></code></pre></div>
<p>The resulting matrix is stored as a binary symmetric matrix of float values, as we can check.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatInfo</span>(tmppeafile1)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; File:               /tmp/RtmpAg94fG/TrapnellPearson.bin</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Matrix type:        SymmetricMatrix</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of elements: 82944 (41616 really stored)</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data type:          float</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Endianness:         little endian (same as this machine)</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of rows:     288</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of columns:  288</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata:           Stored only names of rows.</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Metadata comment:  &quot;Pearson dissimilarities for cell</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          in experiment conquer</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          GSE52529-GPL16791&quot;</span></span></code></pre></div>
</div>
<div id="applying-pam" class="section level2">
<h2>Applying PAM</h2>
<p>The last step is to take the previously calculated matrix and apply the Partitioning Around Medoids classifier. Function is <code>ApplyPAM</code>. Default parameters are:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>L<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;BUILD&quot;</span>,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">max_iter=</span><span class="dv">1000</span>,<span class="at">nthreads=</span><span class="dv">0</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/RtmpAg94fG/TrapnellPearson.bin</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 250891632 Kib.</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting BUILD initialization method, serial version</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 50. TD=11914.134766</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 28. 8 reassigned points. TD=10251.591797</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 284. 79 reassigned points. TD=9409.731445</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 107. 0 reassigned points. TD=8996.875977</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 268. 21 reassigned points. TD=8661.666016</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 118. 0 reassigned points. TD=8333.292969</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 17. 108 reassigned points. TD=8016.939453</span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 154. 0 reassigned points. TD=7747.802734</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 126. 0 reassigned points. TD=7500.490234</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 173. 5 reassigned points. TD=7287.086914</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 7287.086914</span></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; BUILD initialization method (serial version) finished. Elapsed time: 0.00314656 s</span></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting improved FastPAM1 method in serial implementation...</span></span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 0. Medoid at place 1 (point 51) swapped with point 209; TD-change=-103.419434; TD=7183.665039. 60 reassigned points.</span></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 1. Medoid at place 7 (point 17) swapped with point 39; TD-change=-43.391033; TD=7140.273926. 11 reassigned points.</span></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 2.    Exiting, since DeltaTDst is 0.000000. Final value of TD is 7140.273926</span></span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Optimization method (serial version) finished. Elapsed time: 0.00133563 s</span></span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.00314656 s (method BUILD).</span></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0.00133563 s in 1 iterations (0.00133563 seconds/iteration).</span></span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00448219 s (0 minutes, 0.00448219 seconds).</span></span></code></pre></div>
<p>The dissimilarity matrix, as formerly calculated, is compulsory, as long as the number of medoids, <code>k</code>.</p>
<p>Parameters <code>init_method</code> (and one optional parameter, <code>initial_med</code>) deserve special comment. The first is the method to initialize the medoids. Its possible values are <code>BUILD</code>, <code>LAB</code> and <code>PREV</code>. The rest of the algorithm make medoid swapping between the points of the initial set made with <code>BUILD</code> or <code>LAB</code> and the rest of points until no swap can reduce the objective function, which is the sum of distances of each point to its closest medoid. But this may fall (and indeed falls) in local minima. If you initialize with <code>BUILD</code> or <code>LAB</code> the optional parameter <code>initial_med</code> cannot be used.</p>
<p>The initialization methods <code>BUILD</code> and <code>LAB</code> are described in the paper by Schubert at al. (<span class="citation">Schubert and Rousseeuw (2019)</span>). <code>BUILD</code> is deterministic. <code>LAB</code> uses a sample of the total points to initialize. Obviously, you can run <code>LAB</code> to get different initializations and compare the results.</p>
<p>The returned object is a list with two fields: <code>med</code> and <code>clasif</code>. This will be explained later.</p>
<p>From now on, typical calls to obtain only the initial medoids would be</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>Lbuild<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;BUILD&quot;</span>,<span class="at">max_iter=</span><span class="dv">0</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/RtmpAg94fG/TrapnellPearson.bin</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 250891632 Kib.</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting BUILD initialization method, serial version</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 50. TD=11914.134766</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 28. 8 reassigned points. TD=10251.591797</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 284. 79 reassigned points. TD=9409.731445</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 107. 0 reassigned points. TD=8996.875977</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 268. 21 reassigned points. TD=8661.666016</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 118. 0 reassigned points. TD=8333.292969</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 17. 108 reassigned points. TD=8016.939453</span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 154. 0 reassigned points. TD=7747.802734</span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 126. 0 reassigned points. TD=7500.490234</span></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 173. 5 reassigned points. TD=7287.086914</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 7287.086914</span></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; BUILD initialization method (serial version) finished. Elapsed time: 0.00315888 s</span></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.00315888 s (method BUILD).</span></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0 s in 0 iterations.</span></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00315888 s (0 minutes, 0.00315888 seconds).</span></span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>Llab1<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;LAB&quot;</span>,<span class="at">max_iter=</span><span class="dv">0</span>)</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/RtmpAg94fG/TrapnellPearson.bin</span></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 250891632 Kib.</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting LAB initialization method, serial version.</span></span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 38. TD=12058.174805</span></span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 178. 87 reassigned points. TD=11036.697266</span></span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 174. 74 reassigned points. TD=10732.007812</span></span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 190. 22 reassigned points. TD=10437.152344</span></span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 28. 9 reassigned points. TD=8785.389648</span></span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 90. 5 reassigned points. TD=8659.027344</span></span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 225. 33 reassigned points. TD=8541.425781</span></span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 266. 26 reassigned points. TD=8378.083008</span></span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 104. 31 reassigned points. TD=8277.411133</span></span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 16. 5 reassigned points. TD=8124.220703</span></span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 8124.220703</span></span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; LAB initialization method (serial version) finished. Elapsed time: 0.000260456 s</span></span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.000260456 s (method LAB).</span></span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0 s in 0 iterations.</span></span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.000260456 s (0 minutes, 0.000260456 seconds).</span></span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>Llab2<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;LAB&quot;</span>,<span class="at">max_iter=</span><span class="dv">0</span>)</span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/RtmpAg94fG/TrapnellPearson.bin</span></span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb19-57"><a href="#cb19-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 250891632 Kib.</span></span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb19-60"><a href="#cb19-60" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb19-61"><a href="#cb19-61" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb19-62"><a href="#cb19-62" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting LAB initialization method, serial version.</span></span>
<span id="cb19-63"><a href="#cb19-63" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb19-64"><a href="#cb19-64" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 1. Medoid 1 found. Point 171. TD=12271.138672</span></span>
<span id="cb19-65"><a href="#cb19-65" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 2. Medoid 2 found. Point 40. 9 reassigned points. TD=10734.589844</span></span>
<span id="cb19-66"><a href="#cb19-66" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 3. Medoid 3 found. Point 87. 147 reassigned points. TD=9936.333984</span></span>
<span id="cb19-67"><a href="#cb19-67" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 4. Medoid 4 found. Point 284. 74 reassigned points. TD=9271.624023</span></span>
<span id="cb19-68"><a href="#cb19-68" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 5. Medoid 5 found. Point 245. 10 reassigned points. TD=9054.750000</span></span>
<span id="cb19-69"><a href="#cb19-69" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 6. Medoid 6 found. Point 23. 22 reassigned points. TD=8809.705078</span></span>
<span id="cb19-70"><a href="#cb19-70" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 7. Medoid 7 found. Point 174. 50 reassigned points. TD=8623.922852</span></span>
<span id="cb19-71"><a href="#cb19-71" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 8. Medoid 8 found. Point 269. 3 reassigned points. TD=8514.175781</span></span>
<span id="cb19-72"><a href="#cb19-72" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 9. Medoid 9 found. Point 52. 25 reassigned points. TD=8445.003906</span></span>
<span id="cb19-73"><a href="#cb19-73" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Looking for medoid 10. Medoid 10 found. Point 259. 24 reassigned points. TD=8366.524414</span></span>
<span id="cb19-74"><a href="#cb19-74" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Current TD: 8366.524414</span></span>
<span id="cb19-75"><a href="#cb19-75" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; LAB initialization method (serial version) finished. Elapsed time: 0.000264999 s</span></span>
<span id="cb19-76"><a href="#cb19-76" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb19-77"><a href="#cb19-77" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0.000264999 s (method LAB).</span></span>
<span id="cb19-78"><a href="#cb19-78" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0 s in 0 iterations.</span></span>
<span id="cb19-79"><a href="#cb19-79" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.000264999 s (0 minutes, 0.000264999 seconds).</span></span></code></pre></div>
<p>As you can see, to get and compare different initializations you must set the parameter <code>max_iter</code> to value <code>0</code>. In this case no iterations of objective function reduction are performed, and the returned object contains the initial medoids and the classification induced by them. Notice that even looking equal, the results of the latter two calls are different since <code>LAB</code> initializes with a random component (the sample to choose initial medoids is chosen randomly).</p>
<p>You can check that the medoids, stored in <code>Llab1$med</code> and <code>Llab2$med</code> (see more on this below) are in general, different.</p>
<p>Now, these results can be used to initialize PAM if you find that any of them contains a specially good set of medoids. This is the role of method <code>PREV</code> that we have mentioned before. A typical call would be</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>Llab2Final<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafile1,<span class="at">k=</span><span class="dv">10</span>,<span class="at">init_method=</span><span class="st">&quot;PREV&quot;</span>,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">initial_med=</span>Llab2<span class="sc">$</span>med)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/RtmpAg94fG/TrapnellPearson.bin</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 250891632 Kib.</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 288 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting improved FastPAM1 method in serial implementation...</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 0. Medoid at place 1 (point 172) swapped with point 107; TD-change=-321.770386; TD=8044.753418. 27 reassigned points.</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 1. Medoid at place 9 (point 52) swapped with point 118; TD-change=-258.746094; TD=7786.006836. 25 reassigned points.</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 2. Medoid at place 10 (point 259) swapped with point 154; TD-change=-191.391617; TD=7594.615234. 26 reassigned points.</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 3. Medoid at place 8 (point 269) swapped with point 22; TD-change=-168.869598; TD=7425.745605. 6 reassigned points.</span></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 4. Medoid at place 6 (point 23) swapped with point 268; TD-change=-89.157509; TD=7336.588379. 2 reassigned points.</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 5. Medoid at place 2 (point 40) swapped with point 126; TD-change=-86.092148; TD=7250.496582. 5 reassigned points.</span></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 6. Medoid at place 7 (point 174) swapped with point 209; TD-change=-37.088821; TD=7213.407715. 16 reassigned points.</span></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 7. Medoid at place 3 (point 87) swapped with point 39; TD-change=-34.442234; TD=7178.965332. 9 reassigned points.</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 8. Medoid at place 8 (point 22) swapped with point 28; TD-change=-30.032187; TD=7148.933105. 0 reassigned points.</span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 9. Medoid at place 5 (point 245) swapped with point 173; TD-change=-8.662140; TD=7140.270996. 3 reassigned points.</span></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 10.    Exiting, since DeltaTDst is 0.000000. Final value of TD is 7140.270996</span></span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Optimization method (serial version) finished. Elapsed time: 0.00470662 s</span></span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0 s (method PREV).</span></span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0.00470662 s in 9 iterations (0.000522958 seconds/iteration).</span></span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00470662 s (0 minutes, 0.00470662 seconds).</span></span></code></pre></div>
<p>where the initial set of medoids is taken from the object returned by the former calls.</p>
<p>With respect to that object, as we said it is a list with two vectors. The first one, <code>L$med</code>, has as many components as requested medoids and the second, <code>L$clasif</code>, has as many components as instances (here, cells).</p>
<p>Medoids are expressed in <code>L$med</code> by its number in the array of points (row in the dissimilarity matrix) starting at 1 (R convention).</p>
<p><code>L$clasif</code> contains the number of the medoid (i.e.: the cluster) to which each instance has been assigned, according to their order in <code>L$med</code> (also from 1).</p>
<p>This means that if <code>L$clasif[p]</code> is <code>m</code>, the point <code>p</code> belongs to the class grouped around medoid <code>L$med[m]</code>. Let us see it:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Which are the indexes of the points chosen as medoids?</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>L<span class="sc">$</span>med</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1269168 GSM1268987 GSM1269243 GSM1269066 GSM1269227 GSM1269077 GSM1268998 </span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        209         28        284        107        268        118         39 </span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1269113 GSM1269085 GSM1269132 </span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        154        126        173</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co"># In which class has point 147 been classified?</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>L<span class="sc">$</span>clasif[<span class="dv">147</span>]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1269106 </span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;          7</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co"># And which is the index (row in the dissimilarity matrix) of the medoid closest to point 147?</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>L<span class="sc">$</span>med[L<span class="sc">$</span>clasif[<span class="dv">147</span>]]</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; GSM1268998 </span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         39</span></span></code></pre></div>
<p>In this way, values in L$clasif are between 1 and the number of medoids, as we can see.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(L<span class="sc">$</span>clasif)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 1</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(L<span class="sc">$</span>clasif)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 10</span></span></code></pre></div>
<p>Therefore, they can be used as factors.</p>
<p>To help in the data analysis a data frame can be built, too, with the names of the cells instead of their numbers. It is created as</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>F <span class="ot">&lt;-</span> <span class="fu">ClassifAsDataFrame</span>(L,tmppeafile1)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span></code></pre></div>
<p>The second parameter is the dissimilarity matrix created before when we called <code>CalcAndWriteDissimilarityMatrix</code>. The reason of using it here is because the returned dataframe has three columns: <code>CellName</code> with name of each cell, <code>NNCellName</code> with the name of the cell which is the center of the cluster to which <code>CellName</code> belongs to and <code>NNDistance</code> which is the distance between the cells <code>CellName</code> and <code>NNCellName</code>.</p>
<p>This data frame implicitly knows which cells are medoids: those whose distance to the closest medoid (themselves) is 0. They can be extracted as</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract column 1 (cell name) of those rows for which distance to</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># closest medoid (column 3) is 0</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>F[<span class="fu">which</span>(F[,<span class="dv">3</span>]<span class="sc">==</span><span class="dv">0</span>),<span class="dv">1</span>]</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;GSM1268987&quot; &quot;GSM1268998&quot; &quot;GSM1269066&quot; &quot;GSM1269077&quot; &quot;GSM1269085&quot;</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [6] &quot;GSM1269113&quot; &quot;GSM1269132&quot; &quot;GSM1269168&quot; &quot;GSM1269227&quot; &quot;GSM1269243&quot;</span></span></code></pre></div>
<p>Finally, an abundance matrix can be obtained if the cells came from different groups (samples) by combining such groups with the obtained clusters. This is done by function <code>BuildAbundanceMatrix</code>,</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>M<span class="ot">=</span><span class="fu">BuildAbundanceMatrix</span>(L<span class="sc">$</span>clasif,Gr)</span></code></pre></div>
<p>where the <code>Gr</code> vector should have been obtained before (either from the Seurat object with <code>GetSeuratGroups</code>, or by any other means from your input object, if possible). Since the data of this example does not come from Seurat, this is not calculated now, but in a real case the resulting object is a matrix with as many rows as clusters and as many columns as groups. Each entry has the number of cells of its sample which have been classified at that group. This should be analyzed to determine if being in a different group has sufficient influence on the cluster to which cells belong and therefore if these clusters have any real biological significance.</p>
</div>
<div id="alternative-workflow-filtering-by-silhouette" class="section level2">
<h2>Alternative workflow: filtering by silhouette</h2>
<p>It is interesting to filter cells based on the degree in which they belongs to a cluster. Indeed, cluster refinement can be done getting rid of cells far away from any cluster center, or which are at a similar distance of two or more of them.</p>
<p>This is characterized by the silhouette of each point (cell). Three functions deal with this: <code>CalculateSilhouette</code>, <code>FilterBySilhouetteQuantile</code> and <code>FilterBySilhouetteThreshold</code>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>S<span class="ot">=</span><span class="fu">CalculateSilhouette</span>(Llab2<span class="sc">$</span>clasif,tmppeafile1,<span class="at">nthreads=</span><span class="dv">0</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 162 KiB, which is 0% of the available memory, which is 250891632 Kib.</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Calculating silhouette (parallel version) with 128 threads.</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 288 points classified in 10 classes.</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finished parallel implementation of silhouette (including dissimilarity matrix load). Elapsed time: 0.00449829 s</span></span></code></pre></div>
<p>The parameters to function <code>CalculateSilhouette</code> are the array of class membership, as returned by <code>ApplyPAM</code> in its <code>clasif</code> field, and the file with the matrix of dissimilarities.</p>
<p>A parallel implementation has been programmed, being nthreads as explained before.</p>
<p>Silhouette is a number in <span class="math inline">\([-1,1]\)</span>; the higher its value, the most centered a point is in its cluster.</p>
<p>The returned object <code>S</code> is a numeric vector with the value of the silhouette for each point, which will be named if the classification vector was named.</p>
<p>This vector can be converted to object of the class <code>cluster:silhouette</code> with the function <code>NumSilToClusterSil</code> (which needs the vector of classifications, too). This is done so that, if you load the package cluster, plot will generate the kind of silhouette plots included in such package.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>Sclus <span class="ot">&lt;-</span> <span class="fu">NumSilToClusterSil</span>(Llab2<span class="sc">$</span>clasif,S)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Sclus)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2dd1wUx/vHnyvSQVBQBCxYsaNfFL/qV8/YC7FhVzSGmNhQYiOWxBZjS6KmiJqY2DVqNP5MMRpzsaKi2MWIYsGGCqKA9Oc3s7u35dhjJQJyee3nj9mdss/tvm/LzOwzs4CqChS87h0o7VIBKUgFpCAVkIJUQApSASlIBaQgFZCCVEAKUgEpSAWkIBWQglRAClIBKUgFpCAVkIJUQApSASlIBaQgFZCCVEAKUgEpSAWkIBWQglRAClIBKUgFpCAVkIJUQApSASlIBaSgogeUvSEsKHjyj3l0vZNhqjh4eZ3avTtXWBSgfHYtbbJrQIfFfOT38UFB4/cp2qIqckCXfYFRg5skYgddxMHLayBAhrAoQPnsWtjkONmlgdx68kB2F/93T8EWVVEDyqwD8J9RgysBBOQgDhv4SSkBtABg0kl2NasegMavpTNA3fSCbVEVNaBTACPJIvMNgPPcSWsC9GxJnxG/M4XytkzsPm5tFlm7azCsJIuVBsNdGls0oO+8h2Rlel2Ayb+aFkI6I7pN3Mxec2OFgxIMmjYx+x2MALjCpX4KUO8qOY86ACwi0WPj3+w1/phgK/3rd3qEH+f3ragBbQH4lC6PhYaelF5iLQLpab2G5N03MGd4k2uIceRwSMpkgDhyZ3Cnyc4HEPV0JcK0ENIZkW2CvUmCw2bTQYkMcpswEiW3rwLQfCKTmusK2st0JV4DHRFnsZfbJyZbV+vRqOZD074VwxnkPPNIGhcTA4Lyb7cBqEBS+wG0md+FBFJAKR5Q55sVXuDyEP8iJ+DvN7iFKN0ECNze6wHgksIZFxnktmQkSma4GZjU6wCd2PyUp88wVgO1F0x1B5tMzlZbgOHTfUFzqJgA4TC6K1q/kM1ZZoDsr2MeIfQYzwJ0JA+aQQD7JIAWAlxE/ANgjvQeJE7nAGnPIk4E+JI1LjYougeJk1PfBzj9gkn+BSBM2N3tNWv+gTgO4BJr6xDAOMQ7ephQXIDwx34ezElb/54U0P+QOag4/BZgL1k/Sc9rMaCBYBMRETEJaHExIHE6B4hcGngbYBRrXGxQBEiSTO5BsWzyBoCPJDv8947pTgBnWVtfABwhaQd2Hy02QHTff5pEbhsj8j/mGRBkX8k9Ep8CvCUB1Bw4NZUCEqdzgMbQpT29VKhdsUERIEmyAIg88N8W7eu3VekthwdE/sLbyP9OMQC6HhXF3IBulwE/C4DIJXOarN8B/jRm07uAh5HRKSkgcTq348PIIksLQ1i7YoMiQJJkAVAygB9Ti81o3SIE95LY0iMLeEAzAf4uVkCLANbSZW5ZaGUB0C4AWqf9DiASbwD0IeshNH0s2JJDy4uOvi4FJE7ndrwqqWPtB5jN2hUbFAGSJAuAkNwIv6PL7wHG4giABOYexAEiiZvIj3VuMZPbt6IGFKOBspHXH0f3Zx62soCeVgA3I56sBA63MVMLTrsefleGpv9BN8mZA/Atc5h3kFuI0zlAMObJuTqgi2Xtig2atiSSJIsAndaCdsrBY3McQB9Fi+/AEy48oAR7qBWf9gnA19y+Ffk9aLbpjtE41QIgXEdy7YH9fzszZSsw6d0BPMoBtCPPP/LQ8VxmWojSOUAVmK1GI2dXbJDbkkqcLAKES/XcLpJyq5gypBJykLNF998WoOEL074VNaC8bY2p2fIfJqElQLivFinh8wMtHk9uwbazJzHpGWFkzxzC6T0szps5A9mFKJ0DNOl9PdhNyTX9gtggtyUjUbIYEJ5oSYjoWh0mq7kjAfTjVoOwj5Hkv4DgBNO+FcdT7PnFU4ns2iHjeXGA141GtjKSZCqBebfOZPLp2RcusfmYEXMuVVgI6cjdPFNjMoVfkBjkNkFp8g2jUdzwyr54OZNbfRCdjqlGY7Rg61b0M2HfrLA/iHvylZBUQAqyRkB6faE6315NVgioZKUCUpAKSEEqIAWpgBSkAlJQ6QV093Je0RvNOHtF4TWJuUofoBGHaJjYDsD7uFz+RdoBm0/zDr2E6YTuOgDHDxXeRUpV6gAd19F+UuzrHX07yP2FTAF5QD7LlE3fL19r+/Uz07QzC7M/pQzQps56piP5jm4T4mPdd5LMvKu0K4ACemakTc0o2vGTdPJaHuYZPcaR1ibei35OwnRjdtYhLkeskKrJdDHJ/nkh9qiUAdo5M4wBtAUek7BFf3He8SoA1U4xgKJoNyBWW4oYrgdo8iDTT1+hF95qAeBE0mJhn0c1Lke0/V3N18zy+uwbhdijUgaI3CcYQJ850PX+AaKMJy5DHpw3eKRLAP0Eq54ecRjHXGK5DVpeTVkA2wmgenue8zm89sHlf7A/pRTQNC+6PspPlLGgXBriKf+LEkBLbJIQDx1mAO3WUH+JNzoQQAtRyOG1HNKw8CpBQBlhdZ2a7SErj0K9XTvFiFdEYgHNdqfrIeIzaBD3WkwC6JJDpcm/0gc3ATTbfQdRTx8C6CAKObzWwX1+feHSl93rEgTUweWz/wvWHMC8NpW/22VwSxBWxGIBrdTT+2tQkCij4wB2KQCqSo7yRkRDTdnNDKDRzgaqHgQQhW7K4RUNfzLLtGZr8WEivqSKA1C2bOpx+JE8iAI64d+wG/Gp/kthRSwW0EE4R0KfSaKMd1qQIGv2FRYQeYDl2S/Fx+Refqe7fTYF9IkvLZbyhAA6i0IOr3SPYGa5B4yFOJiiBhQHP9cEz+kyOd9q6Ak/rgo5Bf4iR0ceN/yKWCyg3MrhiAdAfPlt15xE3AFxFFAMbEPcCkvxnVqk1ve1TQb6fIanKO/HlSaygPgcQes11PHkRpX6Wa/xEosDj3nGt8mjJJ+Smfd+TQyYV7fdzRthdpeFFbFYQLhV/85095HijLwuZcdNcBrBXGIZlStOnWjvuhQPaZvPGOA4BrHhfzfi6DJDJ/h632UB8TkijdXUDenkVOHM67wHxcFExMzyM+Rzc8dqyOHfcgEA+jKGXxHpkYFtYBwY1X+VtJ6X82X/fqtyMd4QT35m9P8afj9xG/VD6jJ0EzlX9vd9n1Ad0WMWubncNtAKpSlHrL8mBw355Cm+XkDUNzKAfTO1lHtBt5fLPBto8xWp4Qa22HPwbY8TwkrJ63UCoi4GHKCUm6zYN1J5i8q0pS+eIh3pO8X2vYWVktfrBETbRAERMlkRtquZ5cgGNJzmLayUvEoJIMkldsnkfLDInraQ/veGsFLyKiWAJJfYLP3q74l2YZJ3oy3/N0BrFFZKXHnvfvWyRUvuEuO8t0nz6vZgb1cDqQIJKyWtdf0fvWzRUtdYLW1SASlIBaQgFZCCShWg4wmKRe6RFt2Fb1actFjg2PItkkE8S7NEkasrNwqdsE9/+tyyGV6lCtBLvJqInIwfgq8fvCefnTNA19THVfxg9BTczXCN3te9/BEucsnTrT58zK7v18RZ+sFSBeibGMUikZN/geVkAXtksz+1OYY5Y7xEZ40IUJL9VMzu6M9Gcqv0yMYIXTxdf+IF1gHIdIk9Nd4V0hKTD8QIvTqRk8cy/WK+4xKiqEv87Pdp7AztXks5+RBb0W7HVJ2ot0UE6HOb54hH4M+zF4i93yGekOl9kGb0qWolgEyXmBGEXkafD9zLgM8FUzRy8go6cinPozypdLZ/jl5aEstwnoPpIzSgDakVSjOd5gg2RYDeon2SOXp7AE8jjm/Ap3/rs9nKACV8eV5I063OPll2gCkaybonbgKv6Ge7tMtxPe0c2wPXcIznwfTdzrUrk5NkJ7wr2BQB6tKTBI81De9fbdoA+3Q7OjZ4MW0EXXM+8JuVAZKkdSPB8BbM+qOTdxlAOYt1/vRGXG8yZpeby2Qnlll/80ruXDu3muHB5VxFHYmej06bVpsNRczaBF0Rr/yIrau6vzPSxe8FZgdOxL1g8e5X+gHNJsHkFhjV43l/2pQjx37e33ZezqOfPx9KvV3fCsBsty/JVVkRoMoYONbXgQ4y5O7SmR/U1PrXI9sysU69cJ0jgHbY5seI9bV0uATUfjir3ot1dqAZlZX/t6lKPyDaMUEApd2IcNqTdszZD7fZdY7Dxbp6gxf5EUB7NQn79Im4Bsq3aG7naogLc/6/JJ1N2apVqwYhjun72HOeTVoa+655WKvzMO1hGW11G304vhlA7fnZ9tP1HQ31obt+ifw+WQ0gxDq0i6CLwznb6Yg3YfolxOYEUKZL5KjuiOF0bNjasvNDHebh1zDYhfFPuO3+HD1j3I9xplbYjaydZ4QBvs9mwNEP6lF7tZq5t27taV8f/lPPV94rpnQB+lwmjQMU1WM+7bzu6DaSjoX6HSAY450n4e1rQzpX2IzYwo4O2e3sC9DiqXODGIii2+4iJ5FnTIDpEnts4zEOm8MGuPAz/LlV0+sEroL+tQj5ceQm/ZuFN/elCxB3Bh3zXCuk8YAQ06NX2bWr61itWrUqOqeQ8Cq1GkYZ3HeBUxrh0F9Xr2tj/f5eYOeoi5kPa+/ff4zLQyWAyLOvqp9beU9oBxUyjE119fy0b7qQ/8RlMQF0mX+1IFWpBHTWsItP67eNBCvHMAcZBVBt9MLZVOO6dp9653jv9d8vTm8/i1R9dN0c3O10pP7TQTPYzkcPNQIDR+CONymgutEmQC9g6Mb7t96DZh1/QmPwldXTAJrfJ6mr42Yn3YOtsvtUqgBVLKAtRg8yLzXa11XWczEdXC9i7ji723iwUR19859hP029Xi6FAHJeZwKUq6MV0BvMuWIMZuwF1swTp+ZXKQL0eCWss5zLHeQ0uCCb7RROggewiR74+UZfAVvTfLvnI89ZZZJMgLDiPBKcZoayGtkX9ceJPVFqfpUiQAeqDEu3nBvVVk/rhsvhd9lsP/qIS4ENOKMt4uV6OvYxnzqhikMAXw9CbEPfwa3R09dxxo6MvVhiT5SaX6UIUMGK6uZKh2DOLSPvuPJxtWeEnv46DvTKxCfVRFMHxAo3aVztlIC5bXox6xm8PXFqPlkNoE39tum7Tu+lWcTFPUMl2cmNqk/uryHXynnblpNqNRT5JhJAm/px66mBded1cSOttx2e+5G3x6fKyVoAXep0HP+a0v09/iWa4SNpgafze4X+RlfOv91r8RNRRmwA3ZZTyse9JpEaJu6nLhK8PVOqnKwF0D9XbIBymQL0LwJEWvPMMvvcdXGr4Ww9caGsmKf5NmTndTDK93hYGaBd/SzlXK4D0JD6aM23A6gbLWT8UU5UirTmxe32qMEkyOE8CCbK2rUuQHfcfCzkJHr0vnnF10Aq3bAw6eJ/qgrdZOJLjLTmH/1ow7fbc4OYvFiqvWXk/ZSsClBeOwcekFF6V13uRupQa10TMZA0L0itT5jiTgxoVG1SD5/ItduT36oGQl5n+RPIugAt9gvjAUGwJKvlMG6lyWckuAff8zliQHWo572p3Z48c6Y/n7fNR/R6SCxrAhTjeCqCB/Tlr5I8z4+vf/vtdVNsuWhEkBiQC53GQ9RuDzXlZVWNtPCjVgQove48jLBwDyKteWd/H9tv2Nhqncg/VgSItNuRnl58u50HFFk5E+VlRYDGtMixCEhozZN2eQfNRFEHswhQvna7CVBOZYseZ9YD6IRmjdE42MNCdYVvzeMOl4AocY74EjNvt5sA/aI1m5RTkPUA2l1gdYVvzRvLTMmR5IgBmbfbTYAGdLD4s9YDKC+HaKp3jnzfOt+ab2UwyxEDMm+3c4CynSz7dFoPIEbCPchCaz4FAkZQCbU+MSBJax55QEfhqMVftDJAq/mmhoXWfLyB1UE+Q9JYlbTmEZew76i3tZMbPczKygD9A6mteQWpgBQUP/iVNv/3A3pFqYAUpAJSkApIQSogBamAFKQCUpAKSEFWCKgR0+thYVhHTCe3OlPF/R3ioQjJwyvWnJGTL7KoSdnuFmfMsUJATiMjieT72C/ZBu2J9BwmShEDau677nOHMPNIuNOn2/3q/wu6XDklwmHLmQPqk3PiqOaakCIC9Bf1XPzCMVkauUcnu/rDgleNNQKKAnH3qMFdklmbGQXk8pmQIgI0oTIJ7mvWSSNLXWkH9jlL08FYH6DNulEVKgTf4mIfvSXJbNWVBI80U4QUEaBg5ossbtOlkZC2mHjRghc5WiOg+RC8c3VVrxTZzK81kSlXO4LorY8IUGtmxEeNsdJI5849AMpadP6zPkCnfyLBVb18L3LeVB1oJlSTP4NaMR8Yqx4mjTSDd+8mjtdamkHE+gCxajTUQsaz6CR0+VSIiwD1YSb5KveJNNKpWi5ijpeFV/NWCCiGqQE1lR+TmUDfPv8tfiSJb9LVkd6gNkgj5B5EFGipW836ALXyy0a8abNWNnOB3QPEqZVENUURoEPUOXitXZI0ss3lEeJDp8XmpjhZH6DduqBt62r7c5OTzX1HknnLuevPM7SrRSniimLLRr9vcZ2A+GTgRiGS09D/h23+lZ9Z+DnrA4R/vuHmO9r0EDOrB+GhZk4tN4oTxIBShnrWnJ5HpwGbLEQwabCX56C7aEHWA8jkgignM7dEXumnJYCo5JwUmbGMFkxbCyDeBVFOIrfEKMnNNjbAHJCckyIyYxnbP0Q5WQkg3gVRTiK3RM7r0KR8gOSdFJmxjB6DZI1bCSDeBZGRmX8i75Yo9TpEBpC0xi3vpMiMZfzAQ/aXrQRQS3EPhrl/Iu+WKPE6pMr3VlXeSZEZyzipluwvWwkgqQuimX8iI84tMbRgQJacFNOjV7nIjAfF0ghI7nFEXRAbF+hARd0SL33GHnE839KggMSQXkC9sk03C06K8YFc5i5m7KGcSh8gOVEXxLyVhjKfRC7WbpbJZ9wSX9T3YQFN4pGIARmNxjP3oM72MFjJn0GT3LnMSU3p2EO5n7YOQIIL4oTOMrmsW+J4oICarhumlQOU4+np2W2OdhZi38ask+ILUpIB9ILZ5Lj8WEYrAWRyQTzjEs+liG7UrFvir66DKaDGrVtXlD2DqBpX7E1nEWadFO+Tkgygg5qGZJNY+V5XKwFkckFsxc8/GCx0CDFuiYmeGyO4SyzUIiD33k4JGAMtTfFQBlCG62iyiYWxjFYCiHNB/M2Rmf9Q6p/IuiVWrnZCEVCWZllg3XkGWGJyUmQB4TZ95UrCWEaprAQQ54IYyEynZOafyLgl1rZtdZACol6HBQCKTPm41yg6NTrrpBjqwzop/tWorDCWUSprAcToDJy3kHPdcZ/gAmv5Eiu3EGlzgx+hyZeUVp/EsipAExpaypF4mVsG1Jj25u80fcr33weo0ixLOYkHiAa6H6CfBxcdLfVPFAOa65WBGBLIx/9lgC6DqIkxMH/LIP8lRiUGlODR7/AC7U7EwwPPSEqWQkD/4ONhX2qS0hb1Hrye2dSswUoV5hw0lGbyR9uH+odL2mPzPW1q0ar4RlKZTpweVNOX2iJGa/ha2h9rOoMQM5t4hI2weVs+M8mzWliIXjS/G9sXJAYU6vBWT+0as+IFGrUyQFvhEuIauCmbOdfjAZ0EXTRTtDmga5qN5E7vlS0tXqBRKwO0xI4EMSb3jok9JZl96YduDsMZIcUc0Myy2fRO9pO0uNSouawLUBTswMz3Hbhe1J6185eY7mQ+TWmu8Lm1we1o6DRXWlxq1FxWAUi4gS7XNHZ1+tNiwX7+buJvHZq/0XijLw2rhpkVL9CoVQDi9bht3Q8nlR8j/3UlohnDnPqJvuZjDihwCA1rj5EWL9iodQEaUyMd8W/bDZZLnNXsFSIiQONJey2rBzOPUIXZ0uIFG7UWQDcmB025gHWYP7/ZcLPMu/N7D9uCC5mxvJVChHTPWP6hvzsyMjIn1B9TPw7SslPhkuLnxnZzHEqNEuuePVFWVgLoqGPzqW2cj7Smp0CO9xRpZlyVqmHBmnG1aUMrTesqZNApAsX6QXO7oXcQODMdirVHHtI3D9fCltY9iPUptjZHUE5WAqhRp1zE4f9dC8ue3Q+1Occm7uHmnxjmQ2o0n2vCbX5ITwiASsJW5oAyPJvZH23QIaBfauRxXGDToO2t/s7tG6wFrzfuhtq8+V/ZX7YOQElA7xDH4O+FDgBeu7n+IJPjQs0JJHgEG963BdA2EE09YA4IL9hqNS2ffK2/SL0X3tcAVN8/zwU/AmqUWJf7aesAlMzMKWGE/Zhx7u9cdGc9N89yY5XX0+nHjpL6X9bpRtMiCgKEXuMT6McO/zDSd2xZZ2/kZhk6EesfEqNGdu5Fc1kBoP2k5tuscdKOpnqYb6nMFwE6/YJsnO6fGeHz84+mVAKIm6CbU57+C9zRzBlML87+bmCjrbEku1njxIU1dDBeznLpB5RL5/w95+0ANoHAzM0ho3lQznGYftYhh4sY4enAuy/mA/QC1uzVjF4NtpydyVDdxqCfdc7bVmPXEuBnGdOlHdCdr9rQ+wU+re/31wWoKPd1ScQUG+h2EyfZVf0ccaoN8IB+vSQAyiUVoTHotLRdF9Iaq8DayXQJwxW6cfY5T93aEOt18/eglH5Av7RubUePMQm+xZ2aEF82VfxVAHJeNAVa99sBmjYGg4vGuRmfEysAyiMVoR+x5mhiZz8Monbuzo+C3/EazIQb6DGHWG/fA/OrtAMiqkmOsXNXOIYD2i7XsGNOLgWL3RcmlTPS9kW4XcTMmcN0jt7t+JxYs0sMx9QidqZVW0bt3IO1cRl4AIbadxz4nveJ3rXLfCfz61YCaIkGoldotm9ge3siJe9Ac11H0q+8LdOT+mNylc4+LQQX6nyATmvh4jGnBRvgROQt7OR7IvNsw2r6KUs0u5qQR730m6OcrARQdjsAl1W4HpiXxtLu1lvcG422pLk5oGtEgYBwOtho381dD6tgLz7rqAWoCCOyswcB2E9d4Lwa88tKAOF52JCDuMKWSZF6mD03Go0L3Sp9nIi4pfz9G8ctATpAIEYQOynUzkMjnc702TJ33920HEQS6/PKy7jeWAugJ5r1JBxfQ77Iz7oxzPwtE7mTaTeXLgGUFhsbmyi1Y9puPTwm4a9wDfPJWgDhG33IhVZ9mmyJbC/OQ+8qfT1Wt/0BkxtC/ktMYoff7gTTDTvTTmZUlNUA+kU358hgN25o6eWB+8Ql/oCp9AOc37OzAIkusefGfIB4O6sGZgjbdSu/4rcZZebJ/LrVAMLtzcu2N31gRloPwkjuymJnkW4hHgiUDxBvJxRShe3SPqjt6L9G7t1YaQX0D94ryio/oEKqtAIqKqmAFGTNgCyPTSlCZVqevu3l9G8/g15ZKiAFqYAUpAJSkApIQSogBamAFKQCUlCJAMoOX1U4Lf2gkBvMWljIDSbJT46SXyUC6KbjqMKpQ41CblC/dSE34D+dqaQSAXTXq5AbbB2gXEaisV8ql5Go0bmXLKgCUpAKSEEqIAWpgBSkAlKQCkhBKiAFlQigh1UKucHOIYXcYMKqQm7Q1NJXMc1VMm2xx4Usny03CVJBep6hXEail94jtbGqIBWQglRAClIBKUgFpCAVkIJUQApSASlIBaQgFZCCVEAKKnZAz2L4L1YkMd+9v1pQaclUtPLz0r6KeXPdvazo6lfMgLLDtGBncmDfzHhMhhZQXDIVrdm8tEVgntEIfmR9YjsA7+MFlcViB/SFfuujj+AUG5lfM5ZIfj5HRpKpaM3mpS0C84yO6/hx4329o28HuSu4oBUzID86Ybj/cDYysrtCaclUtJJIkZgn2tRZz88yeYd+kuWx7ruCtyheQPdhBwkjKrAxw4THp5IKKi6ZilYSKRLzRDtnhvGAtjDjD1r0L3iL4gV0Bk6S8GtgZ0Co4lcGYFABL8UlU9Gaz0v76uYZJfAWP3OgYX+Lc0+xKl5AvzJDrTezLvCZ2ubnnm50tHwXfcGMbeamopVEisQ8KwHQNKanfJRfweWLDVAyeeTeMDJzGK9l/+LcB/SJPENnsXc0V0e73m+wRyCJWFDhzLMSAM1mht2HvK4z6JCnp+f8K8y0RZ86i9J/gSsWt6lIR5OcZqeilUbkVUjzjARAK/W0EhQUVHD54r3E0uyXk3DIf5jI+aX0n94Klu8SbXqTYA07Fa00UiTmGQmADgJ99eMzqeDyxfyYH9wsGx+WW8asn4U5iDnt5afIYLTaKQFz2/SSiRSJeUYCoNzK4XQYYkyBxYsb0GWPjvMbNCEthvc90zBM135Kk3IF7FBqYN15XdzOslPR8pEiM8+IA9SiMznb9O9Mdx+pUL6422JXp/acR7+Zt8RAaqy73wuabvETKFQpH/eaRN/oMVPRmiJFZ57qETPLLQ6hMwsdGNV/lVJjTG3NK0gFpCAVkIJUQApSASlIBaQgFZCCVEAKUgEpSAWkIBWQglRAClIBKahkAMV/0KdHOO2IuGq4i/GGeLxriFPeKi3LFFgQtcZq7kdMyZcyWziVCKCfyzSaGNZE9w1ijOsNvAgXMQ4K7uhh5LPMFFgQtcZqYDBT8qXMFk4lAsi7J+11ec+O/WBB0QESZN2AHgBzkBf9T2EcuShYQDGR/QYwkyGuCwlelI542UDxDdxGgl9H9fjoEWYZbGuGMoEphSrVsAdxi+FPxNWdc6k1zFjce8yJgcFMSZFZohvvdxuyW2Z/CqWSAJTp5Gd6P3oW4jhArTvNCdBGIYaUGT7Oo2E6RgH9THq1pYif6AaF+/g+yN1aLuQ3JjClMBYajkQcAFMQO3RgrGUHOo4c5lI9mCkpmCW6Va7+xA4g90nEwqhELrHN9lB/wk/08w0CoB55mFFuPh6GHxDjbJcKgG7arEZMLD9euMSEFKL3ayBWrvNfzHFazFhbpTmG+AOYLjHOLNUKmyeIQeGvuO8l8xRLXhfiA07zxYDoud94Js5gpn8O7igAWuFC/RXeqy8AElKQvk1NuAPflkk/BecYa72ZiUl9TYA4s1Tr4BPaXY0HT77KrpdcPehsb9gkAkRfCPrPxMFtaN4HNQVAYbYNiCpUEAAJKURptpu31UjWGj+vmMdYazKapnY1AeLMUmWPtteFQJ8AAAR+SURBVNEHLkzFTXKzIL+0SgLQD+zsszmVQkSA6OOGHMnoJjRrfBMOUMWlOKU64yl2WAAkpFC1e29CCPrP7zuMPR/bMyOnOoifYhwgxJSdoQ6tXnHnSwLQL8B8PzDd8aP8gFbY0q+YNxpEAJ1HvAtLca0jfTe6ZJUASEihWlC3WSSGda64gQU0zjsT8ZmrDKBI6h7yFaQaS/0lltvSLfLyg6Ody8blB5RSvmtC8lQ4iQkwPDmpGwGU5t0+Lukz6tXh8xEygZBCdRL053G7XvOABXRVN+Te3T42wUxJCaCP7fZmXgmqgbMLOxpRohK5B2W86wAAzQ9jfkAY5QsaZ/rNgw+0NtoG/uQxH+MH4LCApAzVtGYDPoUqtxy5Zz8Af2St4Q4XgJ6Tg5mSEkBpXUEDfqesARA5qmvRzPyqqcYXmGZMwxdG6mkQHU+CvOsX2M/q3PtjZ9pZ5vuMN6OZLzrlXfibC0wpjM5Ek+DIZc4auRmfv4/xl5iSIrNUCVFx5PlnFYBep1RAClIBKaj0P8WsWiogBamAFKQCUpAKSEEqIAWpgBSkAlKQCkhBKiAFqYAUpAJSkApIQSogBamAFKQCUlAxAMrtbnjZOfgs6vX5Ef0pHgFEIsUA6ABoZryqjdfmR7Re22Ok01BRpBgAhXr3qVkUdl6HH1Fu1VDEXcxYYS5S9ICy3KZsg2jM6fIN0uFrv/DePbcNScv7YObywd3D6bcE81b2Dbv8Lp0ZQPD+4X16StSP6HtyLXM6QOe5yPMMFyJFD2gPnEu1n4LYKxDpoNxE3rsnFoY1m5fX2m3omGoeyYjDy44d7Oq+DCXePyafnpL0I8LF7vxI6dU66kjS4w0hUvSABtdH7Fs1Dzdo7iN26SZ498RC11y8pt9IL579eAL+QlwAZt4/Jp+ekvQjEmtueRoObyREihxQmtNCOi3GMXxqswafltksePfEwk62SO5PsBen0xkPEsHM+8fk01OSfkRiTfOm4ah6QqTIAW0Ff4MhECYgdgvC9U5pgndPLDOPzZmh9e28CKCBnWlpBzPvH5NPT0n6EYm11ImG/Q1CpMgB9ao2myjAKxfX2qf1HIaCd08sfXScsR+6L+EFAfR2a1I4R2fu/cP59JSsH5GgjUAnvWo1SIgUNaCntnPoYjMY8Yl+k90+FLx7GEBzKiOd1mYvLnVJRXK0Zt4/Jp+ekvQjEuuejjxWn3GfDWciRQ1oLTC16BQbckZ39KqYg4J3DwPoc40R494gz6nk8iOT4/+rXy71/jH59JSkHxFGRwrzX/SqnZA6rBz57zbvZCNFDagTN9tMN48cXAUT6arJu4cB9LwVOJRdG2j7Oxo9NGXWO61CifePyaenJP2IcDZXMaRKaaux8TpDVmoGsJGiBnSYq1LEG9PxuZH75jLr3ZPOOO/g1XNZ+CI6Lflq1tnUZPhFyGfF+vSUqB/RTfJDgu6xM+OdOM1GXltr/jBl82HFwk61XuJ6fd0dY7X1qrv+9Np+/mX1GvuDYrf/mfz6fv1lpXaYKUgFpKD/B96w5WX9uKmHAAAAAElFTkSuQmCC" /><!-- --></p>
<p>Once the silhouette is calculated we can filter it by quantile or by threshold. All points under this quantile or threshold will be discarded, except if they are medoids. Parameters are the silhouette, as returned by <code>CalculateSilhouette</code>, the list of medoids/clasif, as returned by <code>ApplyPAM</code>, the file with matrix of counts for the whole set of cells, the file that will contain the matrix of counts of the remaining cells, the file with dissimilarity matrix for the whole set, the file that will contain the dissimilarity for the remaining cells and (depending on the function used) the quantile in <span class="math inline">\([0,1]\)</span> or the silhouette threshold in <span class="math inline">\([-1,1]\)</span>. As an example:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>tmpfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/Trapnell.bin&quot;</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>tmpfiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellFilt.bin&quot;</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>tmppeafile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearson.bin&quot;</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>tmppeafiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearsonFilt.bin&quot;</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>Lfilt<span class="ot">=</span><span class="fu">FilterBySilhouetteQuantile</span>(S,Llab2,tmpfile1,tmpfiltfile1,</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>                                 tmppeafile1,tmppeafiltfile1,<span class="fl">0.2</span>)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; After filtering silhouette with quantile 0.2 (threshold -0.157013) 233 of the 288 points remain.</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in FilterBySilhouetteQuantile(S, Llab2, tmpfile1, tmpfiltfile1, : 4</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; of the medoids have been kept, even they were below the threshold (which seems</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; problematic. Check your clusters...).</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4 of the medoids have been kept, even they were below the threshold (which seems problematic. Check your clusters...).</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/RtmpAg94fG/TrapnellFilt.bin of (233x5218)</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 2451212</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (233 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing column names (5218 strings written, from ENSG00000164675.10 to ENSG00000093100.13).</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Experiment conquer GSE52529-GPL16791 Filtered by silhouette from file /tmp/RtmpAg94fG/Trapnell.bin with quantile 0.2. </span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (288,288)</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Writing binary matrix /tmp/RtmpAg94fG/TrapnellPearsonFilt.bin</span></span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; End of block of binary data at offset 109172</span></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing row names (233 strings written, from GSM1268960 to GSM1269247).</span></span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Writing comment: Pearson dissimilarities for cell</span></span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          in experiment conquer</span></span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                          GSE52529-GPL16791 Filtered by silhouette from file /tmp/RtmpAg94fG/TrapnellPearson.bin with quantile 0.2.</span></span></code></pre></div>
<p>If the original matrix contained row (in this case, cell) and column (in this case, gene) names, the column names are copied and the row names are transported for those rows (cells) that remain. The same happens with respect to rows of the dissimilarity matrix.</p>
<p>Notice that the new dissimilarity matrix could have been calculated from the matrix of filtered counts with <code>CalcAndWriteDissimilarityMatrix</code> but creating it here, simply getting rid of the filtered rows and columns is much faster.</p>
<p>Also, if a medoid is below the silhouette quantile, it will not be filtered out, but a warning message will be shown, since this is a strange situation that may indicate that some of your clusters are not real but artifacts due to a few outliers that are close to each other.</p>
<p>Of course you can use the filtered cells to build a data frame, as long as the dissimilarity matrix contains cell names,</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>tmppeafiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearsonFilt.bin&quot;</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>Ffilt<span class="ot">=</span><span class="fu">ClassifAsDataFrame</span>(Lfilt,tmppeafiltfile1)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (233,233)</span></span></code></pre></div>
<p>or to recreate a .csv file of counts of the remaining cells with their correct names using <code>BinMatToCsv</code> like this.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>tmpfiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellFilt.bin&quot;</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>tmpfiltcsvfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellFilt.csv&quot;</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">JMatToCsv</span>(tmpfiltfile1,tmpfiltcsvfile1)</span></code></pre></div>
<p>But remember that this was the result of the first step of the PAM algorithm, so probably you will want to make them iterate.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>tmppeafiltfile1<span class="ot">=</span><span class="fu">paste0</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;/TrapnellPearsonFilt.bin&quot;</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>Lfinal<span class="ot">=</span><span class="fu">ApplyPAM</span>(tmppeafiltfile1,<span class="at">k=</span><span class="fu">length</span>(Lfilt<span class="sc">$</span>med),<span class="at">init_method=</span><span class="st">&quot;PREV&quot;</span>,<span class="at">initial_med=</span>Lfilt<span class="sc">$</span>med)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Reading symmetric distance/dissimilarity matrix /tmp/RtmpAg94fG/TrapnellPearsonFilt.bin</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Package memuse is installed. OK.</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Memory used by the matrix: 106 KiB, which is 0% of the available memory, which is 250913108 Kib.</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   That seems OK.</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Read symmetric matrix with size (233,233)</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Matrix is a correct distance/dissimilarity matrix.</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating with a single thread, since you have only 233 vectors and the overhead of using threads would be excessive.</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting improved FastPAM1 method in serial implementation...</span></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span><span class="al">WARNING</span><span class="co">: all successive messages use R-numbering (from 1) for points and medoids. Substract 1 to get the internal C-numbers.</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 0. Medoid at place 10 (point 208) swapped with point 83; TD-change=-445.389740; TD=8053.854004. 3 reassigned points.</span></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 1. Medoid at place 9 (point 41) swapped with point 91; TD-change=-367.995941; TD=7685.858398. 5 reassigned points.</span></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 2. Medoid at place 1 (point 137) swapped with point 16; TD-change=-253.910706; TD=7431.947266. 26 reassigned points.</span></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 3. Medoid at place 8 (point 216) swapped with point 98; TD-change=-195.278137; TD=7236.669434. 2 reassigned points.</span></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 4. Medoid at place 6 (point 17) swapped with point 215; TD-change=-98.056786; TD=7138.612793. 2 reassigned points.</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 5. Medoid at place 2 (point 32) swapped with point 21; TD-change=-32.181549; TD=7106.431152. 0 reassigned points.</span></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 6. Medoid at place 7 (point 139) swapped with point 132; TD-change=-24.359015; TD=7082.072266. 11 reassigned points.</span></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 7. Medoid at place 3 (point 68) swapped with point 31; TD-change=-25.273798; TD=7056.798828. 8 reassigned points.</span></span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 8. Medoid at place 5 (point 198) swapped with point 138; TD-change=-16.466541; TD=7040.332031. 0 reassigned points.</span></span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 9. Medoid at place 1 (point 16) swapped with point 96; TD-change=-2.188439; TD=7038.143555. 6 reassigned points.</span></span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Iteration 10.    Exiting, since DeltaTDst is 0.000000. Final value of TD is 7038.143555</span></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Optimization method (serial version) finished. Elapsed time: 0.00308193 s</span></span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Time summary  (serial implementation).</span></span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Initalization: 0 s (method PREV).</span></span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Optimization:  0.00308193 s in 9 iterations (0.000342437 seconds/iteration).</span></span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Total time:    0.00308193 s (0 minutes, 0.00308193 seconds).</span></span></code></pre></div>
<p>Of course, we might have used simply 10 as number of medoids, <code>k</code>, since this does not change by filtering, but this is to emphasize the fact that <code>ApplyPAM</code> with method <code>PREV</code> requires both parameters to be consistent.</p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-R-jmatrix" class="csl-entry">
Domingo, Juan. 2022a. <em>Jmatrix: Read from/Write in Disks Matrices with Any Data Type in a Binary Format</em>.
</div>
<div id="ref-R-parallelpam" class="csl-entry">
———. 2022b. <em>Parallelpam: Applies the Partitioning-Around-Medoids (PAM) Clustering Algortihm to Big Sets of Data Using Parallel Implementation, If Several Cores Are Available</em>.
</div>
<div id="ref-R-scellpam" class="csl-entry">
———. 2022c. <em>Scellpam: Applying Partitioning Around Medoids to Single Cell with High Number of Cells</em>.
</div>
<div id="ref-R-DuoClustering2018" class="csl-entry">
Duò, Angelo, and Charlotte Soneson. 2021. <em>DuoClustering2018: Data, Clustering Results and Visualization Functions from Duò Et Al (2018)</em>.
</div>
<div id="ref-R-Seurat" class="csl-entry">
Hao, Yuhan, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, et al. 2021. <span>“Integrated Analysis of Multimodal Single-Cell Data.”</span> <em>Cell</em>. <a href="https://doi.org/10.1016/j.cell.2021.04.048">https://doi.org/10.1016/j.cell.2021.04.048</a>.
</div>
<div id="ref-Schubert2019" class="csl-entry">
Schubert, Erich, and Peter J. Rousseeuw. 2019. <span>“Faster k-Medoids Clustering: Improving the <span>PAM</span>, <span>CLARA</span>, and <span>CLARANS</span> Algorithms.”</span> In <em>Similarity Search and Applications</em>, 171–87. Springer International Publishing. <a href="https://doi.org/10.1007/978-3-030-32047-8_16">https://doi.org/10.1007/978-3-030-32047-8_16</a>.
</div>
<div id="ref-R-splatter" class="csl-entry">
Zappia, Luke, Belinda Phipson, and Alicia Oshlack. 2017. <span>“Splatter: Simulation of Single-Cell RNA Sequencing Data.”</span> <em>Genome Biology</em>. <a href="https://doi.org/10.1186/s13059-017-1305-0">https://doi.org/10.1186/s13059-017-1305-0</a>.
</div>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
